// @tldraw/sync-core@5.2.5/dist-esm/lib/SQLiteSyncStorage.mjs
import { transaction } from "@tldraw/state";
import { assert, objectMapEntries, throttle } from "@tldraw/utils";
import {
  computeTombstonePruning,
  DEFAULT_INITIAL_SNAPSHOT,
  MAX_TOMBSTONES
} from "./InMemorySyncStorage.mjs";
import { MicrotaskNotifier } from "./MicrotaskNotifier.mjs";
import {
  convertStoreSnapshotToRoomSnapshot
} from "./TLSyncStorage.mjs";
function migrateSqliteSyncStorage(storage, {
  documentsTable = "documents",
  tombstonesTable = "tombstones",
  metadataTable = "metadata"
} = {}) {
  let migrationVersion = 0;
  try {
    const row = storage.prepare(`SELECT migrationVersion FROM ${metadataTable} LIMIT 1`).all()[0];
    migrationVersion = row?.migrationVersion ?? 0;
  } catch (_e) {
  }
  if (migrationVersion === 0) {
    migrationVersion++;
    storage.exec(`
			CREATE TABLE ${documentsTable} (
				id TEXT PRIMARY KEY,
				state BLOB NOT NULL,
				lastChangedClock INTEGER NOT NULL
			);

			CREATE INDEX idx_${documentsTable}_lastChangedClock ON ${documentsTable}(lastChangedClock);

			CREATE TABLE ${tombstonesTable} (
				id TEXT PRIMARY KEY,
				clock INTEGER NOT NULL
			);
			CREATE INDEX idx_${tombstonesTable}_clock ON ${tombstonesTable}(clock);

			-- This table is used to store the metadata for the sync storage.
			-- There should only be one row in this table.
			CREATE TABLE ${metadataTable} (
			  migrationVersion INTEGER NOT NULL,
				documentClock INTEGER NOT NULL,
				tombstoneHistoryStartsAtClock INTEGER NOT NULL,
				schema TEXT NOT NULL
			);
			
			INSERT INTO ${metadataTable} (migrationVersion, documentClock, tombstoneHistoryStartsAtClock, schema) VALUES (2, 0, 0, '')
		`);
    migrationVersion++;
  }
  if (migrationVersion === 1) {
    migrationVersion++;
    storage.exec(`
			CREATE TABLE ${documentsTable}_new (
				id TEXT PRIMARY KEY,
				state BLOB NOT NULL,
				lastChangedClock INTEGER NOT NULL
			);
			
			INSERT INTO ${documentsTable}_new (id, state, lastChangedClock)
			SELECT id, CAST(state AS BLOB), lastChangedClock FROM ${documentsTable};
			
			DROP TABLE ${documentsTable};
			
			ALTER TABLE ${documentsTable}_new RENAME TO ${documentsTable};
			
			CREATE INDEX idx_${documentsTable}_lastChangedClock ON ${documentsTable}(lastChangedClock);
		`);
  }
  storage.exec(`UPDATE ${metadataTable} SET migrationVersion = ${migrationVersion}`);
}
const textEncoder = new TextEncoder();
const textDecoder = new TextDecoder();
function encodeState(state) {
  return textEncoder.encode(JSON.stringify(state));
}
function decodeState(state) {
  return JSON.parse(textDecoder.decode(state));
}
class SQLiteSyncStorage {
  /**
   * Check if the storage has been initialized (has data in the clock table).
   * Useful for determining whether to load from an external source on first access.
   */
  static hasBeenInitialized(storage) {
    const prefix = storage.config?.tablePrefix ?? "";
    try {
      const schema = storage.prepare(`SELECT schema FROM ${prefix}metadata LIMIT 1`).all()[0]?.schema;
      return !!schema;
    } catch (_e) {
      return false;
    }
  }
  /**
   * Get the current document clock value from storage without fully initializing.
   * Returns null if storage has not been initialized.
   * Useful for comparing storage freshness against external sources.
   */
  static getDocumentClock(storage) {
    const prefix = storage.config?.tablePrefix ?? "";
    try {
      const row = storage.prepare(`SELECT documentClock FROM ${prefix}metadata LIMIT 1`).all()[0];
      if (row && SQLiteSyncStorage.hasBeenInitialized(storage)) {
        return row.documentClock;
      }
      return null;
    } catch (_e) {
      return null;
    }
  }
  // Prepared statements - created once, reused many times
  stmts;
  sql;
  constructor({
    sql,
    snapshot,
    onChange
  }) {
    this.sql = sql;
    const prefix = sql.config?.tablePrefix ?? "";
    const documentsTable = `${prefix}documents`;
    const tombstonesTable = `${prefix}tombstones`;
    const metadataTable = `${prefix}metadata`;
    migrateSqliteSyncStorage(this.sql, { documentsTable, tombstonesTable, metadataTable });
    this.stmts = {
      // Metadata
      getDocumentClock: this.sql.prepare(
        `SELECT documentClock FROM ${metadataTable} LIMIT 1`
      ),
      getTombstoneHistoryStartsAtClock: this.sql.prepare(
        `SELECT tombstoneHistoryStartsAtClock FROM ${metadataTable}`
      ),
      getSchema: this.sql.prepare(`SELECT schema FROM ${metadataTable}`),
      setSchema: this.sql.prepare(`UPDATE ${metadataTable} SET schema = ?`),
      setTombstoneHistoryStartsAtClock: this.sql.prepare(
        `UPDATE ${metadataTable} SET tombstoneHistoryStartsAtClock = ?`
      ),
      incrementDocumentClock: this.sql.prepare(
        `UPDATE ${metadataTable} SET documentClock = documentClock + 1`
      ),
      // Documents
      getDocument: this.sql.prepare(
        `SELECT state FROM ${documentsTable} WHERE id = ?`
      ),
      insertDocument: this.sql.prepare(`INSERT OR REPLACE INTO ${documentsTable} (id, state, lastChangedClock) VALUES (?, ?, ?)`),
      deleteDocument: this.sql.prepare(
        `DELETE FROM ${documentsTable} WHERE id = ?`
      ),
      documentExists: this.sql.prepare(
        `SELECT id FROM ${documentsTable} WHERE id = ?`
      ),
      iterateDocuments: this.sql.prepare(
        `SELECT state, lastChangedClock FROM ${documentsTable}`
      ),
      iterateDocumentEntries: this.sql.prepare(
        `SELECT id, state FROM ${documentsTable}`
      ),
      iterateDocumentKeys: this.sql.prepare(`SELECT id FROM ${documentsTable}`),
      iterateDocumentValues: this.sql.prepare(
        `SELECT state FROM ${documentsTable}`
      ),
      getDocumentsChangedSince: this.sql.prepare(
        `SELECT state FROM ${documentsTable} WHERE lastChangedClock > ?`
      ),
      // Tombstones
      insertTombstone: this.sql.prepare(
        `INSERT OR REPLACE INTO ${tombstonesTable} (id, clock) VALUES (?, ?)`
      ),
      deleteTombstone: this.sql.prepare(
        `DELETE FROM ${tombstonesTable} WHERE id = ?`
      ),
      deleteTombstonesBefore: this.sql.prepare(
        `DELETE FROM ${tombstonesTable} WHERE clock < ?`
      ),
      countTombstones: this.sql.prepare(
        `SELECT count(*) as count FROM ${tombstonesTable}`
      ),
      iterateTombstones: this.sql.prepare(
        `SELECT id, clock FROM ${tombstonesTable} ORDER BY clock ASC`
      ),
      getTombstonesChangedSince: this.sql.prepare(
        `SELECT id FROM ${tombstonesTable} WHERE clock > ?`
      ),
      // Initial setup (only used when loading a snapshot)
      updateMetadata: this.sql.prepare(
        `UPDATE ${metadataTable} SET documentClock = ?, tombstoneHistoryStartsAtClock = ?, schema = ?`
      )
    };
    const hasData = SQLiteSyncStorage.hasBeenInitialized(sql);
    if (snapshot || !hasData) {
      snapshot = convertStoreSnapshotToRoomSnapshot(snapshot ?? DEFAULT_INITIAL_SNAPSHOT);
      const documentClock = snapshot.documentClock ?? snapshot.clock ?? 0;
      const tombstoneHistoryStartsAtClock = snapshot.tombstoneHistoryStartsAtClock ?? documentClock;
      this.sql.exec(`
				DELETE FROM ${documentsTable};
				DELETE FROM ${tombstonesTable};
			`);
      for (const doc of snapshot.documents) {
        this.stmts.insertDocument.run(doc.state.id, encodeState(doc.state), doc.lastChangedClock);
      }
      if (snapshot.tombstones) {
        for (const [id, clock] of objectMapEntries(snapshot.tombstones)) {
          this.stmts.insertTombstone.run(id, clock);
        }
      }
      this.stmts.updateMetadata.run(
        documentClock,
        tombstoneHistoryStartsAtClock,
        JSON.stringify(snapshot.schema)
      );
    }
    if (onChange) {
      this.onChange(onChange);
    }
  }
  notifier = new MicrotaskNotifier();
  onChange(callback) {
    return this.notifier.register(callback);
  }
  transaction(callback, opts) {
    const clockBefore = this.getClock();
    const trackChanges = opts?.emitChanges === "always";
    return this.sql.transaction(() => {
      const txn = new SQLiteSyncStorageTransaction(this, this.stmts);
      let result;
      let changes;
      try {
        result = transaction(() => {
          return callback(txn);
        });
        if (trackChanges) {
          changes = txn.getChangesSince(clockBefore)?.diff;
        }
      } finally {
        txn.close();
      }
      if (typeof result === "object" && result && "then" in result && typeof result.then === "function") {
        throw new Error("Transaction must return a value, not a promise");
      }
      const clockAfter = this.getClock();
      const didChange = clockAfter > clockBefore;
      if (didChange) {
        this.notifier.notify({ id: opts?.id, documentClock: clockAfter });
      }
      return { documentClock: clockAfter, didChange: clockAfter > clockBefore, result, changes };
    });
  }
  getClock() {
    const clockRow = this.stmts.getDocumentClock.all()[0];
    return clockRow?.documentClock ?? 0;
  }
  /** @internal */
  _getTombstoneHistoryStartsAtClock() {
    const clockRow = this.stmts.getTombstoneHistoryStartsAtClock.all()[0];
    return clockRow?.tombstoneHistoryStartsAtClock ?? 0;
  }
  /** @internal */
  _getSchema() {
    const clockRow = this.stmts.getSchema.all()[0];
    assert(clockRow, "Storage not initialized - clock row missing");
    return JSON.parse(clockRow.schema);
  }
  /** @internal */
  _setSchema(schema) {
    this.stmts.setSchema.run(JSON.stringify(schema));
  }
  /** @internal */
  pruneTombstones = throttle(
    () => {
      const tombstoneCount = this.stmts.countTombstones.all()[0].count;
      if (tombstoneCount > MAX_TOMBSTONES) {
        const tombstones = this.stmts.iterateTombstones.all();
        const result = computeTombstonePruning({ tombstones, documentClock: this.getClock() });
        if (result) {
          this.stmts.setTombstoneHistoryStartsAtClock.run(result.newTombstoneHistoryStartsAtClock);
          this.stmts.deleteTombstonesBefore.run(result.newTombstoneHistoryStartsAtClock);
        }
      }
    },
    1e3,
    // prevent this from running synchronously to avoid blocking requests
    { leading: false }
  );
  getSnapshot() {
    return {
      tombstoneHistoryStartsAtClock: this._getTombstoneHistoryStartsAtClock(),
      documentClock: this.getClock(),
      documents: Array.from(this._iterateDocuments()),
      tombstones: Object.fromEntries(this._iterateTombstones()),
      schema: this._getSchema()
    };
  }
  *_iterateDocuments() {
    for (const row of this.stmts.iterateDocuments.iterate()) {
      yield { state: decodeState(row.state), lastChangedClock: row.lastChangedClock };
    }
  }
  *_iterateTombstones() {
    for (const row of this.stmts.iterateTombstones.iterate()) {
      yield [row.id, row.clock];
    }
  }
}
class SQLiteSyncStorageTransaction {
  constructor(storage, stmts) {
    this.storage = storage;
    this.stmts = stmts;
    this._clock = this.storage.getClock();
  }
  storage;
  stmts;
  _clock;
  _closed = false;
  _didIncrementClock = false;
  /** @internal */
  close() {
    this._closed = true;
  }
  assertNotClosed() {
    assert(!this._closed, "Transaction has ended, iterator cannot be consumed");
  }
  getClock() {
    return this._clock;
  }
  getNextClock() {
    if (!this._didIncrementClock) {
      this._didIncrementClock = true;
      this.stmts.incrementDocumentClock.run();
      this._clock = this.storage.getClock();
    }
    return this._clock;
  }
  get(id) {
    this.assertNotClosed();
    const row = this.stmts.getDocument.all(id)[0];
    if (!row) return void 0;
    return decodeState(row.state);
  }
  set(id, record) {
    this.assertNotClosed();
    assert(id === record.id, `Record id mismatch: key does not match record.id`);
    const clock = this.getNextClock();
    this.stmts.deleteTombstone.run(id);
    this.stmts.insertDocument.run(id, encodeState(record), clock);
  }
  delete(id) {
    this.assertNotClosed();
    const exists = this.stmts.documentExists.all(id)[0];
    if (!exists) return;
    const clock = this.getNextClock();
    this.stmts.deleteDocument.run(id);
    this.stmts.insertTombstone.run(id, clock);
    this.storage.pruneTombstones();
  }
  *entries() {
    this.assertNotClosed();
    for (const row of this.stmts.iterateDocumentEntries.iterate()) {
      this.assertNotClosed();
      yield [row.id, decodeState(row.state)];
    }
  }
  *keys() {
    this.assertNotClosed();
    for (const row of this.stmts.iterateDocumentKeys.iterate()) {
      this.assertNotClosed();
      yield row.id;
    }
  }
  *values() {
    this.assertNotClosed();
    for (const row of this.stmts.iterateDocumentValues.iterate()) {
      this.assertNotClosed();
      yield decodeState(row.state);
    }
  }
  getSchema() {
    this.assertNotClosed();
    return this.storage._getSchema();
  }
  setSchema(schema) {
    this.assertNotClosed();
    this.storage._setSchema(schema);
  }
  getChangesSince(sinceClock) {
    this.assertNotClosed();
    const clock = this.storage.getClock();
    if (sinceClock === clock) return void 0;
    if (sinceClock > clock) {
      sinceClock = -1;
    }
    const diff = { puts: {}, deletes: [] };
    const wipeAll = sinceClock < this.storage._getTombstoneHistoryStartsAtClock();
    if (wipeAll) {
      for (const row of this.stmts.iterateDocumentValues.iterate()) {
        const state = decodeState(row.state);
        diff.puts[state.id] = state;
      }
    } else {
      for (const row of this.stmts.getDocumentsChangedSince.iterate(sinceClock)) {
        const state = decodeState(row.state);
        diff.puts[state.id] = state;
      }
      for (const row of this.stmts.getTombstonesChangedSince.iterate(sinceClock)) {
        diff.deletes.push(row.id);
      }
    }
    return { diff, wipeAll };
  }
}
export {
  SQLiteSyncStorage,
  migrateSqliteSyncStorage
};
