// @tldraw/sync-core@5.2.5/dist-esm/lib/TLSocketRoom.mjs
import { createTLSchema } from "@tldraw/tlschema";
import { getOwnProperty, hasOwnProperty, isEqual, structuredClone } from "@tldraw/utils";
import { JsonChunkAssembler } from "./chunk.mjs";
import { DEFAULT_INITIAL_SNAPSHOT, InMemorySyncStorage } from "./InMemorySyncStorage.mjs";
import { RoomSessionState } from "./RoomSession.mjs";
import { ServerSocketAdapter } from "./ServerSocketAdapter.mjs";
import { TLSyncErrorCloseEventReason } from "./TLSyncClient.mjs";
import { TLSyncRoom } from "./TLSyncRoom.mjs";
import {
  convertStoreSnapshotToRoomSnapshot,
  loadSnapshotIntoStorage
} from "./TLSyncStorage.mjs";
function stripPresenceForSnapshot(record) {
  if (record.typeName !== "instance_presence") return record;
  const stripped = { ...record };
  stripped.scribbles = [];
  stripped.chatMessage = "";
  stripped.selectedShapeIds = [];
  stripped.brush = null;
  return stripped;
}
class TLSocketRoom {
  /**
   * Creates a new TLSocketRoom instance for managing collaborative document synchronization.
   *
   * opts - Configuration options for the room
   *   - initialSnapshot - Optional initial document state to load
   *   - schema - Store schema defining record types and validation
   *   - clientTimeout - Milliseconds to wait before disconnecting inactive clients
   *   - log - Optional logger for warnings and errors
   *   - onSessionRemoved - Called when a client session is removed
   *   - onBeforeSendMessage - Called before sending messages to clients
   *   - onAfterReceiveMessage - Called after receiving messages from clients
   *   - onDataChange - Called when document data changes
   *   - onPresenceChange - Called when presence data changes
   */
  constructor(opts) {
    this.opts = opts;
    if (opts.storage && opts.initialSnapshot) {
      throw new Error("Cannot provide both storage and initialSnapshot options");
    }
    const storage = opts.storage ? opts.storage : new InMemorySyncStorage({
      snapshot: convertStoreSnapshotToRoomSnapshot(
        // eslint-disable-next-line @typescript-eslint/no-deprecated
        opts.initialSnapshot ?? DEFAULT_INITIAL_SNAPSHOT
      )
    });
    if ("onDataChange" in opts && opts.onDataChange) {
      this.disposables.add(
        storage.onChange(() => {
          opts.onDataChange?.();
        })
      );
    }
    this.room = new TLSyncRoom({
      onPresenceChange: opts.onPresenceChange,
      schema: opts.schema ?? createTLSchema(),
      log: opts.log,
      storage,
      clientTimeout: opts.clientTimeout
    });
    this.storage = storage;
    this.room.events.on("session_removed", (args) => {
      this.clearSnapshotTimer(args.sessionId);
      this.sessions.delete(args.sessionId);
      if (this.opts.onSessionRemoved) {
        this.opts.onSessionRemoved(this, {
          sessionId: args.sessionId,
          numSessionsRemaining: this.room.sessions.size,
          meta: args.meta
        });
      }
    });
    this.log = "log" in opts ? opts.log : { error: console.error };
  }
  opts;
  room;
  sessions = /* @__PURE__ */ new Map();
  log;
  storage;
  disposables = /* @__PURE__ */ new Set();
  snapshotTimers = /* @__PURE__ */ new Map();
  /**
   * Returns the number of active sessions.
   * Note that this is not the same as the number of connected sockets!
   * Sessions time out a few moments after sockets close, to smooth over network hiccups.
   *
   * @returns the number of active sessions
   */
  getNumActiveSessions() {
    return this.room.sessions.size;
  }
  /**
   * Handles a new client WebSocket connection, creating a session within the room.
   * This should be called whenever a client establishes a WebSocket connection to join
   * the collaborative document.
   *
   * @param opts - Connection options
   *   - sessionId - Unique identifier for the client session (typically from browser tab)
   *   - socket - WebSocket-like object for client communication
   *   - isReadonly - Whether the client can modify the document (defaults to false)
   *   - meta - Additional session metadata (required if SessionMeta is not void)
   *
   * @example
   * ```ts
   * // Handle new WebSocket connection
   * room.handleSocketConnect({
   *   sessionId: 'user-session-abc123',
   *   socket: webSocketConnection,
   *   isReadonly: !userHasEditPermission
   * })
   * ```
   *
   * @example
   * ```ts
   * // With session metadata
   * room.handleSocketConnect({
   *   sessionId: 'session-xyz',
   *   socket: ws,
   *   meta: { userId: 'user-123', name: 'Alice' }
   * })
   * ```
   */
  handleSocketConnect(opts) {
    const { sessionId, socket, isReadonly = false } = opts;
    const handleSocketMessage = (event) => this.handleSocketMessage(sessionId, event.data);
    const handleSocketError = this.handleSocketError.bind(this, sessionId);
    const handleSocketClose = this.handleSocketClose.bind(this, sessionId);
    this.sessions.set(sessionId, {
      assembler: new JsonChunkAssembler(),
      socket,
      unlisten: () => {
        socket.removeEventListener?.("message", handleSocketMessage);
        socket.removeEventListener?.("close", handleSocketClose);
        socket.removeEventListener?.("error", handleSocketError);
      }
    });
    this.room.handleNewSession({
      sessionId,
      isReadonly,
      socket: new ServerSocketAdapter({
        ws: socket,
        onBeforeSendMessage: this.opts.onBeforeSendMessage ? (message, stringified) => this.opts.onBeforeSendMessage({
          sessionId,
          message,
          stringified,
          meta: this.room.sessions.get(sessionId)?.meta
        }) : void 0
      }),
      meta: "meta" in opts ? opts.meta : void 0
    });
    socket.addEventListener?.("message", handleSocketMessage);
    socket.addEventListener?.("close", handleSocketClose);
    socket.addEventListener?.("error", handleSocketError);
  }
  clearSnapshotTimer(sessionId) {
    const t = this.snapshotTimers.get(sessionId);
    if (t) {
      clearTimeout(t);
      this.snapshotTimers.delete(sessionId);
    }
  }
  scheduleDebouncedSnapshot(sessionId) {
    if (!this.opts.onSessionSnapshot) return;
    this.clearSnapshotTimer(sessionId);
    this.snapshotTimers.set(
      sessionId,
      setTimeout(() => {
        this.snapshotTimers.delete(sessionId);
        const snapshot = this.getSessionSnapshot(sessionId);
        if (snapshot) this.opts.onSessionSnapshot(sessionId, snapshot);
      }, 5e3)
    );
  }
  /**
   * Processes a message received from a client WebSocket. Use this method in server
   * environments where WebSocket event listeners cannot be attached directly to socket
   * instances (e.g., Bun.serve, Cloudflare Workers with WebSocket hibernation).
   *
   * The method handles message chunking/reassembly and forwards complete messages
   * to the underlying sync room for processing.
   *
   * @param sessionId - Session identifier matching the one used in handleSocketConnect
   * @param message - Raw message data from the client (string or binary)
   *
   * @example
   * ```ts
   * // In a Bun.serve handler
   * server.upgrade(req, {
   *   data: { sessionId, room },
   *   upgrade(res, req) {
   *     // Connection established
   *   },
   *   message(ws, message) {
   *     const { sessionId, room } = ws.data
   *     room.handleSocketMessage(sessionId, message)
   *   }
   * })
   * ```
   */
  handleSocketMessage(sessionId, message) {
    const assembler = this.sessions.get(sessionId)?.assembler;
    if (!assembler) {
      this.log?.warn?.("Received message from unknown session", sessionId);
      return;
    }
    try {
      const messageString = typeof message === "string" ? message : new TextDecoder().decode(message);
      const res = assembler.handleMessage(messageString);
      if (!res) {
        return;
      }
      if ("data" in res) {
        if (this.opts.onAfterReceiveMessage) {
          const session = this.room.sessions.get(sessionId);
          if (session) {
            this.opts.onAfterReceiveMessage({
              sessionId,
              message: res.data,
              stringified: res.stringified,
              meta: session.meta
            });
          }
        }
        this.room.handleMessage(sessionId, res.data).catch((error) => {
          this.log?.error?.(error);
          this.room.rejectSession(sessionId, TLSyncErrorCloseEventReason.UNKNOWN_ERROR);
        });
        this.room.pruneSessions();
        this.scheduleDebouncedSnapshot(sessionId);
      } else {
        this.log?.error?.("Error assembling message", res.error);
        this.handleSocketError(sessionId);
      }
    } catch (e) {
      this.log?.error?.(e);
      this.room.rejectSession(sessionId, TLSyncErrorCloseEventReason.UNKNOWN_ERROR);
    }
  }
  /**
   * Handles a WebSocket error for the specified session. Use this in server environments
   * where socket event listeners cannot be attached directly. This will initiate cleanup
   * and session removal for the affected client.
   *
   * @param sessionId - Session identifier matching the one used in handleSocketConnect
   *
   * @example
   * ```ts
   * // In a custom WebSocket handler
   * socket.addEventListener('error', () => {
   *   room.handleSocketError(sessionId)
   * })
   * ```
   */
  handleSocketError(sessionId) {
    this.clearSnapshotTimer(sessionId);
    this.room.handleClose(sessionId);
  }
  /**
   * Handles a WebSocket close event for the specified session. Use this in server
   * environments where socket event listeners cannot be attached directly. This will
   * initiate cleanup and session removal for the disconnected client.
   *
   * @param sessionId - Session identifier matching the one used in handleSocketConnect
   *
   * @example
   * ```ts
   * // In a custom WebSocket handler
   * socket.addEventListener('close', () => {
   *   room.handleSocketClose(sessionId)
   * })
   * ```
   */
  handleSocketClose(sessionId) {
    this.clearSnapshotTimer(sessionId);
    this.room.handleClose(sessionId);
  }
  /**
   * Resumes a previously-connected session directly into `Connected` state, bypassing
   * the connect handshake. Use this after server hibernation (e.g., Cloudflare Durable
   * Object hibernation) when WebSocket connections survived but all in-memory state was lost.
   *
   * The session is restored using a {@link SessionStateSnapshot} previously obtained
   * via {@link TLSocketRoom.getSessionSnapshot}. The client is unaware the server restarted and
   * continues sending messages normally.
   *
   * Unlike {@link TLSocketRoom.handleSocketConnect}, this method does NOT attach WebSocket event
   * listeners. In hibernation environments, events are delivered via class methods
   * (e.g., `webSocketMessage`) rather than `addEventListener`.
   *
   * @param opts - Resume options
   *   - sessionId - Unique identifier for the client session
   *   - socket - WebSocket-like object for client communication
   *   - snapshot - Session state snapshot from {@link TLSocketRoom.getSessionSnapshot}
   *   - meta - Additional session metadata (required if SessionMeta is not void)
   *
   * @example
   * ```ts
   * // After Cloudflare DO hibernation wake
   * for (const ws of ctx.getWebSockets()) {
   *   const data = ws.deserializeAttachment()
   *   room.handleSocketResume({
   *     sessionId: data.sessionId,
   *     socket: ws,
   *     snapshot: data.snapshot,
   *   })
   * }
   * ```
   */
  handleSocketResume(opts) {
    const { sessionId, socket, snapshot } = opts;
    this.sessions.set(sessionId, {
      assembler: new JsonChunkAssembler(),
      socket,
      unlisten: () => {
      }
    });
    this.room.handleResumedSession({
      sessionId,
      isReadonly: snapshot.isReadonly,
      serializedSchema: snapshot.serializedSchema,
      presenceId: snapshot.presenceId,
      presenceRecord: snapshot.presenceRecord,
      requiresLegacyRejection: snapshot.requiresLegacyRejection,
      supportsStringAppend: snapshot.supportsStringAppend,
      socket: new ServerSocketAdapter({
        ws: socket,
        onBeforeSendMessage: this.opts.onBeforeSendMessage ? (message, stringified) => this.opts.onBeforeSendMessage({
          sessionId,
          message,
          stringified,
          meta: this.room.sessions.get(sessionId)?.meta
        }) : void 0
      }),
      meta: "meta" in opts ? opts.meta : void 0
    });
  }
  /**
   * Returns a snapshot of a connected session's state that can be persisted and later
   * used with {@link TLSocketRoom.handleSocketResume} to restore the session after hibernation.
   *
   * Returns `null` if the session doesn't exist or isn't in the `Connected` state.
   *
   * @param sessionId - The session to snapshot
   *
   * @example
   * ```ts
   * // Store snapshot in a Cloudflare WebSocket attachment
   * const snapshot = room.getSessionSnapshot(sessionId)
   * if (snapshot) {
   *   ws.serializeAttachment({ sessionId, snapshot })
   * }
   * ```
   */
  getSessionSnapshot(sessionId) {
    const session = this.room.sessions.get(sessionId);
    if (!session || session.state !== RoomSessionState.Connected) {
      return null;
    }
    let presenceRecord = null;
    if (session.presenceId) {
      const record = this.room.presenceStore.get(session.presenceId);
      if (record) {
        presenceRecord = stripPresenceForSnapshot(record);
      }
    }
    return {
      serializedSchema: session.serializedSchema,
      isReadonly: session.isReadonly,
      presenceId: session.presenceId,
      presenceRecord,
      requiresLegacyRejection: session.requiresLegacyRejection,
      supportsStringAppend: session.supportsStringAppend
    };
  }
  /**
   * Returns the current document clock value. The clock is a monotonically increasing
   * integer that increments with each document change, providing a consistent ordering
   * of changes across the distributed system.
   *
   * @returns The current document clock value
   *
   * @example
   * ```ts
   * const clock = room.getCurrentDocumentClock()
   * console.log(`Document is at version ${clock}`)
   * ```
   */
  getCurrentDocumentClock() {
    return this.storage.getClock();
  }
  /**
   * Retrieves a deeply cloned copy of a record from the document store.
   * Returns undefined if the record doesn't exist. The returned record is
   * safe to mutate without affecting the original store data.
   *
   * @param id - Unique identifier of the record to retrieve
   * @returns Deep clone of the record, or undefined if not found
   *
   * @example
   * ```ts
   * const shape = room.getRecord('shape:abc123')
   * if (shape) {
   *   console.log('Shape position:', shape.x, shape.y)
   *   // Safe to modify without affecting store
   *   shape.x = 100
   * }
   * ```
   */
  getRecord(id) {
    return this.storage.transaction((txn) => {
      return structuredClone(txn.get(id));
    }).result;
  }
  /**
   * Returns information about all active sessions in the room. Each session
   * represents a connected client with their current connection status and metadata.
   *
   * @returns Array of session information objects containing:
   *   - sessionId - Unique session identifier
   *   - isConnected - Whether the session has an active WebSocket connection
   *   - isReadonly - Whether the session can modify the document
   *   - meta - Custom session metadata
   *
   * @example
   * ```ts
   * const sessions = room.getSessions()
   * console.log(`Room has ${sessions.length} active sessions`)
   *
   * for (const session of sessions) {
   *   console.log(`${session.sessionId}: ${session.isConnected ? 'online' : 'offline'}`)
   *   if (session.isReadonly) {
   *     console.log('  (read-only access)')
   *   }
   * }
   * ```
   */
  getSessions() {
    return [...this.room.sessions.values()].map((session) => {
      return {
        sessionId: session.sessionId,
        isConnected: session.state === RoomSessionState.Connected,
        isReadonly: session.isReadonly,
        meta: session.meta
      };
    });
  }
  /**
   * Creates a complete snapshot of the current document state, including all records
   * and synchronization metadata. This snapshot can be persisted to storage and used
   * to restore the room state later or revert to a previous version.
   *
   * @returns Complete room snapshot including documents, clock values, and tombstones
   * @deprecated if you need to do this use
   *
   * @example
   * ```ts
   * // Capture current state for persistence
   * const snapshot = room.getCurrentSnapshot()
   * await saveToDatabase(roomId, JSON.stringify(snapshot))
   *
   * // Later, restore from snapshot
   * const savedSnapshot = JSON.parse(await loadFromDatabase(roomId))
   * const newRoom = new TLSocketRoom({ initialSnapshot: savedSnapshot })
   * ```
   */
  getCurrentSnapshot() {
    if (this.storage.getSnapshot) {
      return this.storage.getSnapshot();
    }
    throw new Error("getCurrentSnapshot is not supported for this storage type");
  }
  /**
   * Retrieves all presence records from the document store. Presence records
   * contain ephemeral user state like cursor positions and selections.
   *
   * @returns Object mapping record IDs to presence record data
   * @internal
   */
  getPresenceRecords() {
    const result = {};
    for (const presence of this.room.presenceStore.values()) {
      result[presence.id] = presence;
    }
    return result;
  }
  /**
   * Loads a document snapshot, completely replacing the current room state.
   * This will disconnect all current clients and update the document to match
   * the provided snapshot. Use this for restoring from backups or implementing
   * document versioning.
   *
   * @param snapshot - Room or store snapshot to load
   *
   * @example
   * ```ts
   * // Restore from a saved snapshot
   * const backup = JSON.parse(await loadBackup(roomId))
   * room.loadSnapshot(backup)
   *
   * // All clients will be disconnected and need to reconnect
   * // to see the restored document state
   * ```
   */
  loadSnapshot(snapshot) {
    this.storage.transaction((txn) => {
      loadSnapshotIntoStorage(txn, this.room.schema, snapshot);
    });
  }
  /**
   * Executes a transaction to modify the document store. Changes made within the
   * transaction are atomic and will be synchronized to all connected clients.
   * The transaction provides isolation from concurrent changes until it commits.
   *
   * @param updater - Function that receives store methods to make changes
   *   - store.get(id) - Retrieve a record (safe to mutate, but must call put() to commit)
   *   - store.put(record) - Save a modified record
   *   - store.getAll() - Get all records in the store
   *   - store.delete(id) - Remove a record from the store
   * @returns Promise that resolves when the transaction completes
   *
   * @example
   * ```ts
   * // Update multiple shapes in a single transaction
   * await room.updateStore(store => {
   *   const shape1 = store.get('shape:abc123')
   *   const shape2 = store.get('shape:def456')
   *
   *   if (shape1) {
   *     shape1.x = 100
   *     store.put(shape1)
   *   }
   *
   *   if (shape2) {
   *     shape2.meta.approved = true
   *     store.put(shape2)
   *   }
   * })
   * ```
   *
   * @example
   * ```ts
   * // Async transaction with external API call
   * await room.updateStore(async store => {
   *   const doc = store.get('document:main')
   *   if (doc) {
   *     doc.lastModified = await getCurrentTimestamp()
   *     store.put(doc)
   *   }
   * })
   * ```
   * @deprecated use the storage.transaction method instead
   */
  // eslint-disable-next-line @typescript-eslint/no-deprecated
  async updateStore(updater) {
    if (this.isClosed()) {
      throw new Error("Cannot update store on a closed room");
    }
    const ctx = new StoreUpdateContext(
      // eslint-disable-next-line @typescript-eslint/no-deprecated
      Object.fromEntries(this.getCurrentSnapshot().documents.map((d) => [d.state.id, d.state])),
      this.room.schema
    );
    try {
      await updater(ctx);
    } finally {
      ctx.close();
    }
    this.storage.transaction((txn) => {
      for (const [id, record] of Object.entries(ctx.updates.puts)) {
        txn.set(id, record);
      }
      for (const id of ctx.updates.deletes) {
        txn.delete(id);
      }
    });
  }
  /**
   * Sends a custom message to a specific client session. This allows sending
   * application-specific data that doesn't modify the document state, such as
   * notifications, chat messages, or custom commands.
   *
   * @param sessionId - Target session identifier
   * @param data - Custom payload to send (will be JSON serialized)
   *
   * @example
   * ```ts
   * // Send a notification to a specific user
   * room.sendCustomMessage('session-123', {
   *   type: 'notification',
   *   message: 'Your changes have been saved'
   * })
   *
   * // Send a chat message
   * room.sendCustomMessage('session-456', {
   *   type: 'chat',
   *   from: 'Alice',
   *   text: 'Great work on this design!'
   * })
   * ```
   */
  sendCustomMessage(sessionId, data) {
    this.room.sendCustomMessage(sessionId, data);
  }
  /**
   * Immediately removes a session from the room and closes its WebSocket connection.
   * The client will attempt to reconnect automatically unless a fatal reason is provided.
   *
   * @param sessionId - Session identifier to remove
   * @param fatalReason - Optional fatal error reason that prevents reconnection
   *
   * @example
   * ```ts
   * // Kick a user (they can reconnect)
   * room.closeSession('session-troublemaker')
   *
   * // Permanently ban a user
   * room.closeSession('session-banned', 'PERMISSION_DENIED')
   *
   * // Close session due to inactivity
   * room.closeSession('session-idle', 'TIMEOUT')
   * ```
   */
  closeSession(sessionId, fatalReason) {
    this.room.rejectSession(sessionId, fatalReason);
  }
  /**
   * Closes the room and disconnects all connected clients. This should be called
   * when shutting down the room permanently, such as during server shutdown or
   * when the room is no longer needed. Once closed, the room cannot be reopened.
   *
   * @example
   * ```ts
   * // Clean shutdown when no users remain
   * if (room.getNumActiveSessions() === 0) {
   *   await persistSnapshot(room.getCurrentSnapshot())
   *   room.close()
   * }
   *
   * // Server shutdown
   * process.on('SIGTERM', () => {
   *   for (const room of activeRooms.values()) {
   *     room.close()
   *   }
   * })
   * ```
   */
  close() {
    this.room.close();
    for (const sessionId of this.snapshotTimers.keys()) {
      this.clearSnapshotTimer(sessionId);
    }
    this.disposables.forEach((d) => d());
    this.disposables.clear();
  }
  /**
   * Checks whether the room has been permanently closed. Closed rooms cannot
   * accept new connections or process further changes.
   *
   * @returns True if the room is closed, false if still active
   *
   * @example
   * ```ts
   * if (room.isClosed()) {
   *   console.log('Room has been shut down')
   *   // Create a new room or redirect users
   * } else {
   *   // Room is still accepting connections
   *   room.handleSocketConnect({ sessionId, socket })
   * }
   * ```
   */
  isClosed() {
    return this.room.isClosed();
  }
}
class StoreUpdateContext {
  constructor(snapshot, schema) {
    this.snapshot = snapshot;
    this.schema = schema;
  }
  snapshot;
  schema;
  updates = {
    puts: {},
    deletes: /* @__PURE__ */ new Set()
  };
  put(record) {
    if (this._isClosed) throw new Error("StoreUpdateContext is closed");
    const recordType = getOwnProperty(this.schema.types, record.typeName);
    if (!recordType) {
      throw new Error(`Missing definition for record type ${record.typeName}`);
    }
    const recordBefore = this.snapshot[record.id] ?? void 0;
    recordType.validate(record, recordBefore);
    if (record.id in this.snapshot && isEqual(this.snapshot[record.id], record)) {
      delete this.updates.puts[record.id];
    } else {
      this.updates.puts[record.id] = structuredClone(record);
    }
    this.updates.deletes.delete(record.id);
  }
  delete(recordOrId) {
    if (this._isClosed) throw new Error("StoreUpdateContext is closed");
    const id = typeof recordOrId === "string" ? recordOrId : recordOrId.id;
    delete this.updates.puts[id];
    if (this.snapshot[id]) {
      this.updates.deletes.add(id);
    }
  }
  get(id) {
    if (this._isClosed) throw new Error("StoreUpdateContext is closed");
    if (hasOwnProperty(this.updates.puts, id)) {
      return structuredClone(this.updates.puts[id]);
    }
    if (this.updates.deletes.has(id)) {
      return null;
    }
    return structuredClone(this.snapshot[id] ?? null);
  }
  getAll() {
    if (this._isClosed) throw new Error("StoreUpdateContext is closed");
    const result = Object.values(this.updates.puts);
    for (const [id, record] of Object.entries(this.snapshot)) {
      if (!this.updates.deletes.has(id) && !hasOwnProperty(this.updates.puts, id)) {
        result.push(record);
      }
    }
    return structuredClone(result);
  }
  _isClosed = false;
  close() {
    this._isClosed = true;
  }
}
export {
  TLSocketRoom
};
