// @tldraw/sync-core@5.2.5/dist-esm/lib/TLSyncStorage.mjs
import { assert, isEqual, objectMapEntriesIterable, objectMapValues } from "@tldraw/utils";
import { diffRecord, RecordOpType } from "./diff.mjs";
function toNetworkDiff(diff) {
  const networkDiff = {};
  for (const [id, put] of objectMapEntriesIterable(diff.puts)) {
    if (Array.isArray(put)) {
      const patch = diffRecord(put[0], put[1]);
      if (patch) {
        networkDiff[id] = [RecordOpType.Patch, patch];
      }
    } else {
      networkDiff[id] = [RecordOpType.Put, put];
    }
  }
  for (const id of diff.deletes) {
    networkDiff[id] = [RecordOpType.Remove];
  }
  return networkDiff;
}
function loadSnapshotIntoStorage(txn, schema, snapshot) {
  snapshot = convertStoreSnapshotToRoomSnapshot(snapshot);
  assert(snapshot.schema, "Schema is required");
  const docIds = /* @__PURE__ */ new Set();
  for (const doc of snapshot.documents) {
    docIds.add(doc.state.id);
    const existing = txn.get(doc.state.id);
    if (isEqual(existing, doc.state)) continue;
    txn.set(doc.state.id, doc.state);
  }
  for (const id of txn.keys()) {
    if (!docIds.has(id)) {
      txn.delete(id);
    }
  }
  txn.setSchema(snapshot.schema);
  schema.migrateStorage(txn);
}
function convertStoreSnapshotToRoomSnapshot(snapshot) {
  if ("documents" in snapshot) return snapshot;
  return {
    clock: 0,
    documentClock: 0,
    documents: objectMapValues(snapshot.store).map((state) => ({
      state,
      lastChangedClock: 0
    })),
    schema: snapshot.schema,
    tombstones: {}
  };
}
export {
  convertStoreSnapshotToRoomSnapshot,
  loadSnapshotIntoStorage,
  toNetworkDiff
};
