// @tldraw/sync-core@5.2.5/dist-esm/index.mjs
import { registerTldrawLibraryVersion } from "@tldraw/utils";
import { chunk, JsonChunkAssembler } from "./lib/chunk.mjs";
import { ClientWebSocketAdapter, ReconnectManager } from "./lib/ClientWebSocketAdapter.mjs";
import {
  applyObjectDiff,
  diffRecord,
  getNetworkDiff,
  RecordOpType,
  ValueOpType
} from "./lib/diff.mjs";
import { DurableObjectSqliteSyncWrapper } from "./lib/DurableObjectSqliteSyncWrapper.mjs";
import { DEFAULT_INITIAL_SNAPSHOT, InMemorySyncStorage } from "./lib/InMemorySyncStorage.mjs";
import { NodeSqliteWrapper } from "./lib/NodeSqliteWrapper.mjs";
import {
  getTlsyncProtocolVersion,
  TLIncompatibilityReason
} from "./lib/protocol.mjs";
import { RoomSessionState } from "./lib/RoomSession.mjs";
import {
  SQLiteSyncStorage
} from "./lib/SQLiteSyncStorage.mjs";
import { TLRemoteSyncError } from "./lib/TLRemoteSyncError.mjs";
import {
  TLSocketRoom
} from "./lib/TLSocketRoom.mjs";
import {
  TLSyncClient,
  TLSyncErrorCloseEventCode,
  TLSyncErrorCloseEventReason
} from "./lib/TLSyncClient.mjs";
import {
  TLSyncRoom
} from "./lib/TLSyncRoom.mjs";
import {
  loadSnapshotIntoStorage
} from "./lib/TLSyncStorage.mjs";
registerTldrawLibraryVersion(
  "@tldraw/sync-core",
  "5.2.5",
  "esm"
);
export {
  ClientWebSocketAdapter,
  DEFAULT_INITIAL_SNAPSHOT,
  DurableObjectSqliteSyncWrapper,
  InMemorySyncStorage,
  JsonChunkAssembler,
  NodeSqliteWrapper,
  ReconnectManager,
  RecordOpType,
  RoomSessionState,
  SQLiteSyncStorage,
  TLIncompatibilityReason,
  TLRemoteSyncError,
  TLSocketRoom,
  TLSyncClient,
  TLSyncErrorCloseEventCode,
  TLSyncErrorCloseEventReason,
  TLSyncRoom,
  ValueOpType,
  applyObjectDiff,
  chunk,
  diffRecord,
  getNetworkDiff,
  getTlsyncProtocolVersion,
  loadSnapshotIntoStorage
};
