// @tldraw/sync-core@5.2.5/dist-cjs/lib/DurableObjectSqliteSyncWrapper.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var DurableObjectSqliteSyncWrapper_exports = {};
__export(DurableObjectSqliteSyncWrapper_exports, {
  DurableObjectSqliteSyncWrapper: () => DurableObjectSqliteSyncWrapper
});
module.exports = __toCommonJS(DurableObjectSqliteSyncWrapper_exports);
class DurableObjectStatement {
  constructor(sql, query) {
    this.sql = sql;
    this.query = query;
  }
  sql;
  query;
  iterate(...bindings) {
    const result = this.sql.exec(this.query, ...bindings);
    return result[Symbol.iterator]();
  }
  all(...bindings) {
    return this.sql.exec(this.query, ...bindings).toArray();
  }
  run(...bindings) {
    this.sql.exec(this.query, ...bindings);
  }
}
class DurableObjectSqliteSyncWrapper {
  constructor(storage, config) {
    this.storage = storage;
    this.config = config;
  }
  storage;
  config;
  exec(sql) {
    this.storage.sql.exec(sql);
  }
  prepare(sql) {
    return new DurableObjectStatement(this.storage.sql, sql);
  }
  transaction(callback) {
    return this.storage.transactionSync(callback);
  }
}
