// @tldraw/sync-core@5.2.5/dist-cjs/lib/InMemorySyncStorage.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var InMemorySyncStorage_exports = {};
__export(InMemorySyncStorage_exports, {
  DEFAULT_INITIAL_SNAPSHOT: () => DEFAULT_INITIAL_SNAPSHOT,
  InMemorySyncStorage: () => InMemorySyncStorage,
  MAX_TOMBSTONES: () => MAX_TOMBSTONES,
  TOMBSTONE_PRUNE_BUFFER_SIZE: () => TOMBSTONE_PRUNE_BUFFER_SIZE,
  computeTombstonePruning: () => computeTombstonePruning
});
module.exports = __toCommonJS(InMemorySyncStorage_exports);
var import_state = require("@tldraw/state");
var import_store = require("@tldraw/store");
var import_tlschema = require("@tldraw/tlschema");
var import_utils = require("@tldraw/utils");
var import_MicrotaskNotifier = require("./MicrotaskNotifier");
const TOMBSTONE_PRUNE_BUFFER_SIZE = 1e3;
const MAX_TOMBSTONES = 5e3;
function computeTombstonePruning({
  tombstones,
  documentClock,
  maxTombstones = MAX_TOMBSTONES,
  pruneBufferSize = TOMBSTONE_PRUNE_BUFFER_SIZE
}) {
  if (tombstones.length <= maxTombstones) {
    return null;
  }
  let cutoff = pruneBufferSize + tombstones.length - maxTombstones;
  while (cutoff < tombstones.length && tombstones[cutoff - 1]?.clock === tombstones[cutoff]?.clock) {
    cutoff++;
  }
  const oldestRemaining = tombstones[cutoff];
  const newTombstoneHistoryStartsAtClock = oldestRemaining?.clock ?? documentClock;
  const idsToDelete = tombstones.slice(0, cutoff).map((t) => t.id);
  return { newTombstoneHistoryStartsAtClock, idsToDelete };
}
const DEFAULT_INITIAL_SNAPSHOT = {
  documentClock: 0,
  tombstoneHistoryStartsAtClock: 0,
  schema: (0, import_tlschema.createTLSchema)().serialize(),
  documents: [
    {
      state: import_tlschema.DocumentRecordType.create({ id: import_tlschema.TLDOCUMENT_ID }),
      lastChangedClock: 0
    },
    {
      state: import_tlschema.PageRecordType.create({
        id: "page:page",
        name: "Page 1",
        index: "a1"
      }),
      lastChangedClock: 0
    }
  ]
};
class InMemorySyncStorage {
  /** @internal */
  documents;
  /** @internal */
  tombstones;
  /** @internal */
  schema;
  /** @internal */
  documentClock;
  /** @internal */
  tombstoneHistoryStartsAtClock;
  notifier = new import_MicrotaskNotifier.MicrotaskNotifier();
  onChange(callback) {
    return this.notifier.register(callback);
  }
  constructor({
    snapshot = DEFAULT_INITIAL_SNAPSHOT,
    onChange
  } = {}) {
    const maxClockValue = Math.max(
      0,
      ...Object.values(snapshot.tombstones ?? {}),
      ...Object.values(snapshot.documents.map((d) => d.lastChangedClock))
    );
    this.documents = new import_store.AtomMap(
      "room documents",
      snapshot.documents.map((d) => [
        d.state.id,
        { state: (0, import_store.devFreeze)(d.state), lastChangedClock: d.lastChangedClock }
      ])
    );
    const documentClock = Math.max(maxClockValue, snapshot.documentClock ?? snapshot.clock ?? 0);
    this.documentClock = (0, import_state.atom)("document clock", documentClock);
    const tombstoneHistoryStartsAtClock = Math.min(
      snapshot.tombstoneHistoryStartsAtClock ?? documentClock,
      documentClock
    );
    this.tombstoneHistoryStartsAtClock = (0, import_state.atom)(
      "tombstone history starts at clock",
      tombstoneHistoryStartsAtClock
    );
    this.schema = (0, import_state.atom)("schema", snapshot.schema ?? (0, import_tlschema.createTLSchema)().serializeEarliestVersion());
    this.tombstones = new import_store.AtomMap(
      "room tombstones",
      // If the tombstone history starts now (or we didn't have the
      // tombstoneHistoryStartsAtClock) then there are no tombstones
      tombstoneHistoryStartsAtClock === documentClock ? [] : (0, import_utils.objectMapEntries)(snapshot.tombstones ?? {})
    );
    if (onChange) {
      this.onChange(onChange);
    }
  }
  transaction(callback, opts) {
    const clockBefore = this.documentClock.get();
    const trackChanges = opts?.emitChanges === "always";
    const txn = new InMemorySyncStorageTransaction(this);
    let result;
    let changes;
    try {
      result = (0, import_state.transaction)(() => {
        return callback(txn);
      });
      if (trackChanges) {
        changes = txn.getChangesSince(clockBefore)?.diff;
      }
    } catch (error) {
      console.error("Error in transaction", error);
      throw error;
    } finally {
      txn.close();
    }
    if (typeof result === "object" && result && "then" in result && typeof result.then === "function") {
      const err = new Error("Transaction must return a value, not a promise");
      console.error(err);
      throw err;
    }
    const clockAfter = this.documentClock.get();
    const didChange = clockAfter > clockBefore;
    if (didChange) {
      this.notifier.notify({ id: opts?.id, documentClock: clockAfter });
    }
    return { documentClock: clockAfter, didChange: clockAfter > clockBefore, result, changes };
  }
  getClock() {
    return this.documentClock.get();
  }
  /** @internal */
  pruneTombstones = (0, import_utils.throttle)(
    () => {
      if (this.tombstones.size > MAX_TOMBSTONES) {
        const tombstones = Array.from(this.tombstones.entries()).map(([id, clock]) => ({ id, clock })).sort((a, b) => a.clock - b.clock);
        const result = computeTombstonePruning({
          tombstones,
          documentClock: this.documentClock.get()
        });
        if (result) {
          this.tombstoneHistoryStartsAtClock.set(result.newTombstoneHistoryStartsAtClock);
          this.tombstones.deleteMany(result.idsToDelete);
        }
      }
    },
    1e3,
    // prevent this from running synchronously to avoid blocking requests
    { leading: false }
  );
  getSnapshot() {
    return {
      tombstoneHistoryStartsAtClock: this.tombstoneHistoryStartsAtClock.get(),
      documentClock: this.documentClock.get(),
      documents: Array.from(this.documents.values()),
      tombstones: Object.fromEntries(this.tombstones.entries()),
      schema: this.schema.get()
    };
  }
}
class InMemorySyncStorageTransaction {
  constructor(storage) {
    this.storage = storage;
    this._clock = this.storage.documentClock.get();
  }
  storage;
  _clock;
  _closed = false;
  /** @internal */
  close() {
    this._closed = true;
  }
  assertNotClosed() {
    (0, import_utils.assert)(!this._closed, "Transaction has ended, iterator cannot be consumed");
  }
  getClock() {
    return this._clock;
  }
  didIncrementClock = false;
  getNextClock() {
    if (!this.didIncrementClock) {
      this.didIncrementClock = true;
      this._clock = this.storage.documentClock.set(this.storage.documentClock.get() + 1);
    }
    return this._clock;
  }
  get(id) {
    this.assertNotClosed();
    return this.storage.documents.get(id)?.state;
  }
  set(id, record) {
    this.assertNotClosed();
    (0, import_utils.assert)(id === record.id, `Record id mismatch: key does not match record.id`);
    const clock = this.getNextClock();
    if (this.storage.tombstones.has(id)) {
      this.storage.tombstones.delete(id);
    }
    this.storage.documents.set(id, {
      state: (0, import_store.devFreeze)(record),
      lastChangedClock: clock
    });
  }
  delete(id) {
    this.assertNotClosed();
    if (!this.storage.documents.has(id)) return;
    const clock = this.getNextClock();
    this.storage.documents.delete(id);
    this.storage.tombstones.set(id, clock);
    this.storage.pruneTombstones();
  }
  *entries() {
    this.assertNotClosed();
    for (const [id, record] of this.storage.documents.entries()) {
      this.assertNotClosed();
      yield [id, record.state];
    }
  }
  *keys() {
    this.assertNotClosed();
    for (const key of this.storage.documents.keys()) {
      this.assertNotClosed();
      yield key;
    }
  }
  *values() {
    this.assertNotClosed();
    for (const record of this.storage.documents.values()) {
      this.assertNotClosed();
      yield record.state;
    }
  }
  getSchema() {
    this.assertNotClosed();
    return this.storage.schema.get();
  }
  setSchema(schema) {
    this.assertNotClosed();
    this.storage.schema.set(schema);
  }
  getChangesSince(sinceClock) {
    this.assertNotClosed();
    const clock = this.storage.documentClock.get();
    if (sinceClock === clock) return void 0;
    if (sinceClock > clock) {
      sinceClock = -1;
    }
    const diff = { puts: {}, deletes: [] };
    const wipeAll = sinceClock < this.storage.tombstoneHistoryStartsAtClock.get();
    for (const doc of this.storage.documents.values()) {
      if (wipeAll || doc.lastChangedClock > sinceClock) {
        diff.puts[doc.state.id] = doc.state;
      }
    }
    if (!wipeAll) {
      for (const [id, clock2] of this.storage.tombstones.entries()) {
        if (clock2 > sinceClock) {
          diff.deletes.push(id);
        }
      }
    }
    return { diff, wipeAll };
  }
}
