// @tldraw/sync-core@5.2.5/dist-cjs/lib/NodeSqliteWrapper.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var NodeSqliteWrapper_exports = {};
__export(NodeSqliteWrapper_exports, {
  NodeSqliteWrapper: () => NodeSqliteWrapper
});
module.exports = __toCommonJS(NodeSqliteWrapper_exports);
class NodeSqliteWrapper {
  constructor(db, config) {
    this.db = db;
    this.config = config;
  }
  db;
  config;
  exec(sql) {
    this.db.exec(sql);
  }
  prepare(sql) {
    return this.db.prepare(sql);
  }
  transaction(callback) {
    this.db.exec("BEGIN");
    let result;
    try {
      result = callback();
    } catch (e) {
      this.db.exec("ROLLBACK");
      throw e;
    }
    this.db.exec("COMMIT");
    return result;
  }
}
