// @tldraw/sync-core@5.2.5/dist-cjs/lib/TLSyncRoom.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var TLSyncRoom_exports = {};
__export(TLSyncRoom_exports, {
  DATA_MESSAGE_DEBOUNCE_INTERVAL: () => DATA_MESSAGE_DEBOUNCE_INTERVAL,
  PresenceStore: () => PresenceStore,
  TLSyncRoom: () => TLSyncRoom
});
module.exports = __toCommonJS(TLSyncRoom_exports);
var import_store = require("@tldraw/store");
var import_utils = require("@tldraw/utils");
var import_nanoevents = require("nanoevents");
var import_diff = require("./diff");
var import_interval = require("./interval");
var import_protocol = require("./protocol");
var import_recordDiff = require("./recordDiff");
var import_RoomSession = require("./RoomSession");
var import_TLSyncClient = require("./TLSyncClient");
var import_TLSyncStorage = require("./TLSyncStorage");
const DATA_MESSAGE_DEBOUNCE_INTERVAL = 1e3 / 60;
const timeSince = (time) => Date.now() - time;
class TLSyncRoom {
  // A table of connected clients
  sessions = /* @__PURE__ */ new Map();
  lastDocumentClock = 0;
  pruneTimer = null;
  pruneSessions = (0, import_utils.throttle)(() => {
    if (this.pruneTimer) {
      clearTimeout(this.pruneTimer);
      this.pruneTimer = null;
    }
    for (const client of this.sessions.values()) {
      switch (client.state) {
        case import_RoomSession.RoomSessionState.Connected: {
          const hasTimedOut = timeSince(client.lastInteractionTime) > this.sessionIdleTimeout;
          if (hasTimedOut || !client.socket.isOpen) {
            this.cancelSession(client.sessionId);
          }
          break;
        }
        case import_RoomSession.RoomSessionState.AwaitingConnectMessage: {
          const hasTimedOut = timeSince(client.sessionStartTime) > import_RoomSession.SESSION_START_WAIT_TIME;
          if (hasTimedOut || !client.socket.isOpen) {
            this.removeSession(client.sessionId);
          } else {
            this.scheduleFollowUpPrune();
          }
          break;
        }
        case import_RoomSession.RoomSessionState.AwaitingRemoval: {
          const hasTimedOut = timeSince(client.cancellationTime) > import_RoomSession.SESSION_REMOVAL_WAIT_TIME;
          if (hasTimedOut) {
            this.removeSession(client.sessionId);
          } else {
            this.scheduleFollowUpPrune();
          }
          break;
        }
        default: {
          (0, import_utils.exhaustiveSwitchError)(client);
        }
      }
    }
  }, 1e3);
  scheduleFollowUpPrune() {
    if (this.pruneTimer) return;
    this.pruneTimer = setTimeout(this.pruneSessions, import_RoomSession.SESSION_REMOVAL_WAIT_TIME + 100);
  }
  presenceStore = new PresenceStore();
  disposables = [];
  _isClosed = false;
  /**
   * Close the room and clean up all resources. Disconnects all sessions
   * and stops background processes.
   */
  close() {
    this.disposables.forEach((d) => d());
    this.sessions.forEach((session) => {
      session.socket.close();
    });
    this._isClosed = true;
  }
  /**
   * Check if the room has been closed and is no longer accepting connections.
   *
   * @returns True if the room is closed
   */
  isClosed() {
    return this._isClosed;
  }
  events = (0, import_nanoevents.createNanoEvents)();
  // Storage layer for documents, tombstones, and clocks
  storage;
  serializedSchema;
  documentTypes;
  presenceType;
  log;
  schema;
  sessionIdleTimeout;
  constructor(opts) {
    this.schema = opts.schema;
    this.log = opts.log;
    this.onPresenceChange = opts.onPresenceChange;
    this.storage = opts.storage;
    this.sessionIdleTimeout = opts.clientTimeout ?? import_RoomSession.SESSION_IDLE_TIMEOUT;
    (0, import_utils.assert)(
      import_utils.isNativeStructuredClone,
      "TLSyncRoom is supposed to run either on Cloudflare Workersor on a 18+ version of Node.js, which both support the native structuredClone API"
    );
    this.serializedSchema = JSON.parse(JSON.stringify(this.schema.serialize()));
    this.documentTypes = new Set(
      Object.values(this.schema.types).filter((t) => t.scope === "document").map((t) => t.typeName)
    );
    const presenceTypes = new Set(
      Object.values(this.schema.types).filter((t) => t.scope === "presence")
    );
    if (presenceTypes.size > 1) {
      throw new Error(
        `TLSyncRoom: exactly zero or one presence type is expected, but found ${presenceTypes.size}`
      );
    }
    this.presenceType = presenceTypes.values().next()?.value ?? null;
    const { documentClock } = this.storage.transaction((txn) => {
      this.schema.migrateStorage(txn);
    });
    this.lastDocumentClock = documentClock;
    this.disposables.push(
      this.storage.onChange(({ id }) => {
        if (id !== this.internalTxnId) {
          this.broadcastExternalStorageChanges();
        }
      })
    );
    this.disposables.push(() => {
      this.pruneSessions.cancel();
      if (this.pruneTimer) {
        clearTimeout(this.pruneTimer);
        this.pruneTimer = null;
      }
    });
    if (Number.isFinite(this.sessionIdleTimeout) && this.sessionIdleTimeout > 0) {
      const pruneIntervalMs = Math.min(2e3, Math.floor(this.sessionIdleTimeout / 4));
      this.disposables.push((0, import_interval.interval)(() => this.pruneSessions(), pruneIntervalMs));
    }
  }
  broadcastExternalStorageChanges() {
    this.storage.transaction((txn) => {
      this.broadcastChanges(txn);
      this.lastDocumentClock = txn.getClock();
    });
  }
  /**
   * Send a message to a particular client. Debounces data events
   *
   * @param sessionId - The id of the session to send the message to.
   * @param message - The message to send. UNSAFE Any diffs must have been downgraded already if necessary
   */
  _unsafe_sendMessage(sessionId, message) {
    const session = this.sessions.get(sessionId);
    if (!session) {
      this.log?.warn?.("Tried to send message to unknown session", message.type);
      return;
    }
    if (session.state !== import_RoomSession.RoomSessionState.Connected) {
      this.log?.warn?.("Tried to send message to disconnected client", message.type);
      return;
    }
    if (session.socket.isOpen) {
      if (message.type !== "patch" && message.type !== "push_result") {
        if (message.type !== "pong") {
          this._flushDataMessages(sessionId);
        }
        session.socket.sendMessage(message);
      } else {
        if (session.debounceTimer === null) {
          session.socket.sendMessage({ type: "data", data: [message] });
          session.debounceTimer = setTimeout(
            () => this._flushDataMessages(sessionId),
            DATA_MESSAGE_DEBOUNCE_INTERVAL
          );
        } else {
          session.outstandingDataMessages.push(message);
        }
      }
    } else {
      this.cancelSession(session.sessionId);
    }
  }
  // needs to accept sessionId and not a session because the session might be dead by the time
  // the timer fires
  _flushDataMessages(sessionId) {
    const session = this.sessions.get(sessionId);
    if (!session || session.state !== import_RoomSession.RoomSessionState.Connected) {
      return;
    }
    session.debounceTimer = null;
    if (session.outstandingDataMessages.length > 0) {
      const data = session.outstandingDataMessages;
      session.outstandingDataMessages = [];
      session.socket.sendMessage({ type: "data", data });
    }
  }
  /** @internal */
  removeSession(sessionId, fatalReason) {
    const session = this.sessions.get(sessionId);
    if (!session) {
      this.log?.warn?.("Tried to remove unknown session");
      return;
    }
    this.sessions.delete(sessionId);
    try {
      if (fatalReason) {
        session.socket.close(import_TLSyncClient.TLSyncErrorCloseEventCode, fatalReason);
      } else {
        session.socket.close();
      }
    } catch {
    }
    const presence = this.presenceStore.get(session.presenceId ?? "");
    if (presence) {
      this.presenceStore.delete(session.presenceId);
      this.broadcastPatch({
        puts: {},
        deletes: [session.presenceId]
      });
    }
    this.events.emit("session_removed", { sessionId, meta: session.meta });
    if (this.sessions.size === 0) {
      this.events.emit("room_became_empty");
    }
  }
  cancelSession(sessionId) {
    const session = this.sessions.get(sessionId);
    if (!session) {
      return;
    }
    if (session.state === import_RoomSession.RoomSessionState.AwaitingRemoval) {
      this.log?.warn?.("Tried to cancel session that is already awaiting removal");
      return;
    }
    this.sessions.set(sessionId, {
      state: import_RoomSession.RoomSessionState.AwaitingRemoval,
      sessionId,
      presenceId: session.presenceId,
      socket: session.socket,
      cancellationTime: Date.now(),
      meta: session.meta,
      isReadonly: session.isReadonly,
      requiresLegacyRejection: session.requiresLegacyRejection,
      supportsStringAppend: session.supportsStringAppend
    });
    try {
      session.socket.close();
    } catch {
    }
    this.scheduleFollowUpPrune();
  }
  internalTxnId = "TLSyncRoom.txn";
  /**
   * Broadcast a patch to all connected clients except the one with the sessionId provided.
   *
   * @param diff - The TLSyncForwardDiff with full records (used for migration)
   * @param networkDiff - Optional pre-computed NetworkDiff for sessions not needing migration.
   *                      If not provided, will be computed from recordsDiff.
   * @param sourceSessionId - Optional session ID to exclude from the broadcast
   */
  broadcastPatch(diff, networkDiff, sourceSessionId) {
    const unmigrated = networkDiff ?? (0, import_TLSyncStorage.toNetworkDiff)(diff);
    if (!unmigrated) return this;
    this.sessions.forEach((session) => {
      if (session.state !== import_RoomSession.RoomSessionState.Connected) return;
      if (sourceSessionId === session.sessionId) return;
      if (!session.socket.isOpen) {
        this.cancelSession(session.sessionId);
        return;
      }
      const diffResult = this.migrateDiffOrRejectSession(
        session.sessionId,
        session.serializedSchema,
        session.requiresDownMigrations,
        diff
      );
      if (!diffResult.ok) return;
      this._unsafe_sendMessage(session.sessionId, {
        type: "patch",
        diff: diffResult.value,
        serverClock: this.lastDocumentClock
      });
    });
    return this;
  }
  /**
   * Send a custom message to a connected client. Useful for application-specific
   * communication that doesn't involve document synchronization.
   *
   * @param sessionId - The ID of the session to send the message to
   * @param data - The custom payload to send (will be JSON serialized)
   * @example
   * ```ts
   * // Send a custom notification
   * room.sendCustomMessage('user-123', {
   *   type: 'notification',
   *   message: 'Document saved successfully'
   * })
   *
   * // Send user-specific data
   * room.sendCustomMessage('user-456', {
   *   type: 'user_permissions',
   *   canEdit: true,
   *   canDelete: false
   * })
   * ```
   */
  sendCustomMessage(sessionId, data) {
    this._unsafe_sendMessage(sessionId, { type: "custom", data });
  }
  /**
   * Register a new client session with the room. The session will be in an awaiting
   * state until it sends a connect message with protocol handshake.
   *
   * @param opts - Session configuration
   *   - sessionId - Unique identifier for this session
   *   - socket - WebSocket adapter for communication
   *   - meta - Application-specific metadata for this session
   *   - isReadonly - Whether this session can modify documents
   * @returns This room instance for method chaining
   * @example
   * ```ts
   * room.handleNewSession({
   *   sessionId: crypto.randomUUID(),
   *   socket: new WebSocketAdapter(ws),
   *   meta: { userId: '123', name: 'Alice', avatar: 'url' },
   *   isReadonly: !hasEditPermission
   * })
   * ```
   *
   * @internal
   */
  handleNewSession(opts) {
    const { sessionId, socket, meta, isReadonly } = opts;
    const existing = this.sessions.get(sessionId);
    this.sessions.set(sessionId, {
      state: import_RoomSession.RoomSessionState.AwaitingConnectMessage,
      sessionId,
      socket,
      presenceId: existing?.presenceId ?? this.presenceType?.createId() ?? null,
      sessionStartTime: Date.now(),
      meta,
      isReadonly: isReadonly ?? false,
      // this gets set later during handleConnectMessage
      requiresLegacyRejection: false,
      supportsStringAppend: true
    });
    return this;
  }
  /**
   * Resume a previously-connected session directly into `Connected` state, bypassing the
   * connect handshake. Used after server hibernation when the WebSocket is still alive but
   * all in-memory state has been lost.
   *
   * @internal
   */
  handleResumedSession(opts) {
    const {
      sessionId,
      socket,
      meta,
      isReadonly,
      serializedSchema,
      presenceId,
      presenceRecord,
      requiresLegacyRejection,
      supportsStringAppend
    } = opts;
    const migrations = this.schema.getMigrationsSince(serializedSchema);
    const requiresDownMigrations = migrations.ok ? migrations.value.length > 0 : false;
    this.sessions.set(sessionId, {
      state: import_RoomSession.RoomSessionState.Connected,
      sessionId,
      socket,
      presenceId: presenceId ?? this.presenceType?.createId() ?? null,
      serializedSchema,
      requiresDownMigrations,
      lastInteractionTime: Date.now(),
      debounceTimer: null,
      outstandingDataMessages: [],
      meta,
      isReadonly,
      requiresLegacyRejection,
      supportsStringAppend
    });
    if (presenceRecord && presenceId) {
      this.presenceStore.set(presenceId, presenceRecord);
    }
  }
  /**
   * Checks if all connected sessions support string append operations (protocol version 8+).
   * If any client is on an older version, returns false to enable legacy append mode.
   *
   * @returns True if all connected sessions are on protocol version 8 or higher
   */
  getCanEmitStringAppend() {
    for (const session of this.sessions.values()) {
      if (session.state === import_RoomSession.RoomSessionState.Connected) {
        if (!session.supportsStringAppend) {
          return false;
        }
      }
    }
    return true;
  }
  /**
   * When we send a diff to a client, if that client is on a lower version than us, we need to make
   * the diff compatible with their version. This method takes a TLSyncForwardDiff (which has full
   * records) and migrates all records down to the client's schema version, returning a NetworkDiff.
   *
   * For updates (entries with [before, after] tuples), both records are migrated and a patch is
   * computed from the migrated versions, preserving efficient patch semantics even across versions.
   *
   * If a migration fails, the session will be rejected.
   *
   * @param sessionId - The session ID (for rejection on migration failure)
   * @param serializedSchema - The client's schema to migrate to
   * @param requiresDownMigrations - Whether the client needs down migrations
   * @param diff - The TLSyncForwardDiff containing full records to migrate
   * @param unmigrated - Optional pre-computed NetworkDiff for when no migration is needed
   * @returns A NetworkDiff with migrated records, or a migration failure
   */
  migrateDiffOrRejectSession(sessionId, serializedSchema, requiresDownMigrations, diff, unmigrated) {
    if (!requiresDownMigrations) {
      return import_utils.Result.ok(unmigrated ?? (0, import_TLSyncStorage.toNetworkDiff)(diff) ?? {});
    }
    const result = {};
    for (const [id, put] of (0, import_utils.objectMapEntriesIterable)(diff.puts)) {
      if (Array.isArray(put)) {
        const [from, to] = put;
        const fromResult = this.schema.migratePersistedRecord(from, serializedSchema, "down");
        if (fromResult.type === "error") {
          this.rejectSession(sessionId, import_TLSyncClient.TLSyncErrorCloseEventReason.CLIENT_TOO_OLD);
          return import_utils.Result.err(fromResult.reason);
        }
        const toResult = this.schema.migratePersistedRecord(to, serializedSchema, "down");
        if (toResult.type === "error") {
          this.rejectSession(sessionId, import_TLSyncClient.TLSyncErrorCloseEventReason.CLIENT_TOO_OLD);
          return import_utils.Result.err(toResult.reason);
        }
        const patch = (0, import_diff.diffRecord)(fromResult.value, toResult.value);
        if (patch) {
          result[id] = [import_diff.RecordOpType.Patch, patch];
        }
      } else {
        const migrationResult = this.schema.migratePersistedRecord(put, serializedSchema, "down");
        if (migrationResult.type === "error") {
          this.rejectSession(sessionId, import_TLSyncClient.TLSyncErrorCloseEventReason.CLIENT_TOO_OLD);
          return import_utils.Result.err(migrationResult.reason);
        }
        result[id] = [import_diff.RecordOpType.Put, migrationResult.value];
      }
    }
    for (const id of diff.deletes) {
      result[id] = [import_diff.RecordOpType.Remove];
    }
    return import_utils.Result.ok(result);
  }
  /**
   * Process an incoming message from a client session. Handles connection requests,
   * data synchronization pushes, and ping/pong for connection health.
   *
   * @param sessionId - The ID of the session that sent the message
   * @param message - The client message to process
   * @example
   * ```ts
   * // Typically called by WebSocket message handlers
   * websocket.onMessage((data) => {
   *   const message = JSON.parse(data)
   *   room.handleMessage(sessionId, message)
   * })
   * ```
   */
  async handleMessage(sessionId, message) {
    const session = this.sessions.get(sessionId);
    if (!session) {
      this.log?.warn?.("Received message from unknown session");
      return;
    }
    try {
      switch (message.type) {
        case "connect": {
          return this.handleConnectRequest(session, message);
        }
        case "push": {
          return this.handlePushRequest(session, message);
        }
        case "ping": {
          if (session.state === import_RoomSession.RoomSessionState.Connected) {
            session.lastInteractionTime = Date.now();
          }
          return this._unsafe_sendMessage(session.sessionId, { type: "pong" });
        }
        default: {
          (0, import_utils.exhaustiveSwitchError)(message);
        }
      }
    } catch (e) {
      if (e instanceof import_TLSyncClient.TLSyncError) {
        this.rejectSession(session.sessionId, e.reason);
      } else {
        throw e;
      }
    }
  }
  /**
   * Reject and disconnect a session due to incompatibility or other fatal errors.
   * Sends appropriate error messages before closing the connection.
   *
   * @param sessionId - The session to reject
   * @param fatalReason - The reason for rejection (optional)
   * @example
   * ```ts
   * // Reject due to version mismatch
   * room.rejectSession('user-123', TLSyncErrorCloseEventReason.CLIENT_TOO_OLD)
   *
   * // Reject due to permission issue
   * room.rejectSession('user-456', 'Insufficient permissions')
   * ```
   */
  rejectSession(sessionId, fatalReason) {
    const session = this.sessions.get(sessionId);
    if (!session) return;
    if (!fatalReason) {
      this.removeSession(sessionId);
      return;
    }
    if (session.requiresLegacyRejection) {
      try {
        if (session.socket.isOpen) {
          let legacyReason;
          switch (fatalReason) {
            case import_TLSyncClient.TLSyncErrorCloseEventReason.CLIENT_TOO_OLD:
              legacyReason = import_protocol.TLIncompatibilityReason.ClientTooOld;
              break;
            case import_TLSyncClient.TLSyncErrorCloseEventReason.SERVER_TOO_OLD:
              legacyReason = import_protocol.TLIncompatibilityReason.ServerTooOld;
              break;
            case import_TLSyncClient.TLSyncErrorCloseEventReason.INVALID_RECORD:
              legacyReason = import_protocol.TLIncompatibilityReason.InvalidRecord;
              break;
            default:
              legacyReason = import_protocol.TLIncompatibilityReason.InvalidOperation;
              break;
          }
          session.socket.sendMessage({
            type: "incompatibility_error",
            reason: legacyReason
          });
        }
      } catch {
      } finally {
        this.removeSession(sessionId);
      }
    } else {
      this.removeSession(sessionId, fatalReason);
    }
  }
  forceAllReconnect() {
    for (const session of this.sessions.values()) {
      this.removeSession(session.sessionId);
    }
  }
  broadcastChanges(txn) {
    const changes = txn.getChangesSince(this.lastDocumentClock);
    if (!changes) return;
    const { wipeAll, diff } = changes;
    this.lastDocumentClock = txn.getClock();
    if (wipeAll) {
      this.forceAllReconnect();
      return;
    }
    this.broadcastPatch(diff);
  }
  /**
   * Work out whether a client we can't reconcile schemas with is running a newer or older SDK
   * than us.
   */
  getVersionMismatchReason(theirSchema) {
    const ourSchema = this.serializedSchema;
    if (theirSchema.schemaVersion > ourSchema.schemaVersion) {
      return import_TLSyncClient.TLSyncErrorCloseEventReason.SERVER_TOO_OLD;
    }
    if (theirSchema.schemaVersion === 2 && ourSchema.schemaVersion === 2) {
      for (const [sequenceId, theirVersion] of Object.entries(theirSchema.sequences)) {
        const ourVersion = ourSchema.sequences[sequenceId];
        if (ourVersion === void 0 || theirVersion > ourVersion) {
          return import_TLSyncClient.TLSyncErrorCloseEventReason.SERVER_TOO_OLD;
        }
      }
    }
    return import_TLSyncClient.TLSyncErrorCloseEventReason.CLIENT_TOO_OLD;
  }
  handleConnectRequest(session, message) {
    let theirProtocolVersion = message.protocolVersion;
    if (theirProtocolVersion === 5) {
      theirProtocolVersion = 6;
    }
    session.requiresLegacyRejection = theirProtocolVersion === 6;
    if (theirProtocolVersion === 6) {
      theirProtocolVersion++;
    }
    if (theirProtocolVersion === 7) {
      theirProtocolVersion++;
      session.supportsStringAppend = false;
    }
    if (theirProtocolVersion == null || theirProtocolVersion < (0, import_protocol.getTlsyncProtocolVersion)()) {
      this.rejectSession(session.sessionId, import_TLSyncClient.TLSyncErrorCloseEventReason.CLIENT_TOO_OLD);
      return;
    } else if (theirProtocolVersion > (0, import_protocol.getTlsyncProtocolVersion)()) {
      this.rejectSession(session.sessionId, import_TLSyncClient.TLSyncErrorCloseEventReason.SERVER_TOO_OLD);
      return;
    }
    if (message.schema == null) {
      this.rejectSession(session.sessionId, import_TLSyncClient.TLSyncErrorCloseEventReason.CLIENT_TOO_OLD);
      return;
    }
    const migrations = this.schema.getMigrationsSince(message.schema);
    if (!migrations.ok) {
      this.rejectSession(session.sessionId, this.getVersionMismatchReason(message.schema));
      return;
    }
    if (migrations.value.some((m) => m.scope !== "record" || !m.down)) {
      this.rejectSession(session.sessionId, import_TLSyncClient.TLSyncErrorCloseEventReason.CLIENT_TOO_OLD);
      return;
    }
    const sessionSchema = (0, import_utils.isEqual)(message.schema, this.serializedSchema) ? this.serializedSchema : message.schema;
    const requiresDownMigrations = migrations.value.length > 0;
    const connect = async (msg) => {
      this.sessions.set(session.sessionId, {
        state: import_RoomSession.RoomSessionState.Connected,
        sessionId: session.sessionId,
        presenceId: session.presenceId,
        socket: session.socket,
        serializedSchema: sessionSchema,
        requiresDownMigrations,
        lastInteractionTime: Date.now(),
        debounceTimer: null,
        outstandingDataMessages: [],
        supportsStringAppend: session.supportsStringAppend,
        meta: session.meta,
        isReadonly: session.isReadonly,
        requiresLegacyRejection: session.requiresLegacyRejection
      });
      this._unsafe_sendMessage(session.sessionId, msg);
    };
    const { documentClock, result } = this.storage.transaction((txn) => {
      this.broadcastChanges(txn);
      const docChanges = txn.getChangesSince(message.lastServerClock);
      const presenceDiff = this.migrateDiffOrRejectSession(
        session.sessionId,
        sessionSchema,
        requiresDownMigrations,
        {
          // Exclude the connecting session's own presence — it will push fresh
          // data immediately after connecting. Sending the stale record back
          // would leave an orphaned presence in the client's local store (the
          // server never echoes a session's own updates back to it).
          puts: Object.fromEntries(
            [...this.presenceStore.values()].filter((p) => p.id !== session.presenceId).map((p) => [p.id, p])
          ),
          deletes: []
        }
      );
      if (!presenceDiff.ok) return null;
      let docDiff = null;
      if (docChanges && sessionSchema !== this.serializedSchema) {
        const migrated = this.migrateDiffOrRejectSession(
          session.sessionId,
          sessionSchema,
          requiresDownMigrations,
          docChanges.diff
        );
        if (!migrated.ok) return null;
        docDiff = migrated.value;
      } else if (docChanges) {
        docDiff = (0, import_TLSyncStorage.toNetworkDiff)(docChanges.diff);
      }
      return {
        type: "connect",
        connectRequestId: message.connectRequestId,
        hydrationType: docChanges?.wipeAll ? "wipe_all" : "wipe_presence",
        protocolVersion: (0, import_protocol.getTlsyncProtocolVersion)(),
        schema: this.schema.serialize(),
        serverClock: txn.getClock(),
        diff: { ...presenceDiff.value, ...docDiff },
        isReadonly: session.isReadonly
      };
    });
    this.lastDocumentClock = documentClock;
    if (result) {
      connect(result);
    }
  }
  handlePushRequest(session, message) {
    if (session && session.state !== import_RoomSession.RoomSessionState.Connected) {
      return;
    }
    if (session) {
      session.lastInteractionTime = Date.now();
    }
    const legacyAppendMode = !this.getCanEmitStringAppend();
    const propagateOp = (changes2, id, op, before, after) => {
      if (!changes2.diffs) changes2.diffs = { networkDiff: {}, diff: { puts: {}, deletes: [] } };
      changes2.diffs.networkDiff[id] = op;
      switch (op[0]) {
        case import_diff.RecordOpType.Put:
          changes2.diffs.diff.puts[id] = op[1];
          break;
        case import_diff.RecordOpType.Patch:
          (0, import_utils.assert)(before && after, "before and after are required for patches");
          changes2.diffs.diff.puts[id] = [before, after];
          break;
        case import_diff.RecordOpType.Remove:
          changes2.diffs.diff.deletes.push(id);
          break;
        default:
          (0, import_utils.exhaustiveSwitchError)(op[0]);
      }
    };
    const addDocument = (storage, changes2, id, _state) => {
      const res = session ? this.schema.migratePersistedRecord(_state, session.serializedSchema, "up") : { type: "success", value: _state };
      if (res.type === "error") {
        throw new import_TLSyncClient.TLSyncError(res.reason, import_TLSyncClient.TLSyncErrorCloseEventReason.CLIENT_TOO_OLD);
      }
      const { value: state } = res;
      const doc = storage.get(id);
      if (doc) {
        const recordType = (0, import_utils.assertExists)((0, import_utils.getOwnProperty)(this.schema.types, doc.typeName));
        const diff = (0, import_recordDiff.diffAndValidateRecord)(doc, state, recordType);
        if (diff) {
          storage.set(id, state);
          propagateOp(changes2, id, [import_diff.RecordOpType.Patch, diff], doc, state);
        }
      } else {
        const recordType = (0, import_utils.assertExists)((0, import_utils.getOwnProperty)(this.schema.types, state.typeName));
        (0, import_recordDiff.validateRecord)(state, recordType);
        storage.set(id, state);
        propagateOp(changes2, id, [import_diff.RecordOpType.Put, state], void 0, void 0);
      }
      return import_utils.Result.ok(void 0);
    };
    const patchDocument = (storage, changes2, id, patch) => {
      const doc = storage.get(id);
      if (!doc) return;
      const recordType = (0, import_utils.assertExists)((0, import_utils.getOwnProperty)(this.schema.types, doc.typeName));
      const downgraded = session ? this.schema.migratePersistedRecord(doc, session.serializedSchema, "down") : { type: "success", value: doc };
      if (downgraded.type === "error") {
        throw new import_TLSyncClient.TLSyncError(downgraded.reason, import_TLSyncClient.TLSyncErrorCloseEventReason.CLIENT_TOO_OLD);
      }
      if (downgraded.value === doc) {
        const diff = (0, import_recordDiff.applyAndDiffRecord)(doc, patch, recordType, legacyAppendMode);
        if (diff) {
          storage.set(id, diff[1]);
          propagateOp(changes2, id, [import_diff.RecordOpType.Patch, diff[0]], doc, diff[1]);
        }
      } else {
        const patched = (0, import_diff.applyObjectDiff)(downgraded.value, patch);
        const upgraded = session ? this.schema.migratePersistedRecord(patched, session.serializedSchema, "up") : { type: "success", value: patched };
        if (upgraded.type === "error") {
          throw new import_TLSyncClient.TLSyncError(upgraded.reason, import_TLSyncClient.TLSyncErrorCloseEventReason.CLIENT_TOO_OLD);
        }
        const diff = (0, import_recordDiff.diffAndValidateRecord)(doc, upgraded.value, recordType, legacyAppendMode);
        if (diff) {
          storage.set(id, upgraded.value);
          propagateOp(changes2, id, [import_diff.RecordOpType.Patch, diff], doc, upgraded.value);
        }
      }
    };
    const { result, documentClock, changes } = this.storage.transaction(
      (txn) => {
        this.broadcastChanges(txn);
        const docChanges = { diffs: null };
        const presenceChanges = { diffs: null };
        if (this.presenceType && session?.presenceId && "presence" in message && message.presence) {
          if (!session) throw new Error("session is required for presence pushes");
          const id = session.presenceId;
          const [type, val] = message.presence;
          const { typeName } = this.presenceType;
          switch (type) {
            case import_diff.RecordOpType.Put: {
              addDocument(this.presenceStore, presenceChanges, id, {
                ...val,
                id,
                typeName
              });
              break;
            }
            case import_diff.RecordOpType.Patch: {
              patchDocument(this.presenceStore, presenceChanges, id, {
                ...val,
                id: [import_diff.ValueOpType.Put, id],
                typeName: [import_diff.ValueOpType.Put, typeName]
              });
              break;
            }
          }
        }
        if (message.diff && !session?.isReadonly) {
          for (const [id, op] of (0, import_utils.objectMapEntriesIterable)(message.diff)) {
            switch (op[0]) {
              case import_diff.RecordOpType.Put: {
                if (!this.documentTypes.has(op[1].typeName)) {
                  throw new import_TLSyncClient.TLSyncError(
                    "invalid record",
                    import_TLSyncClient.TLSyncErrorCloseEventReason.INVALID_RECORD
                  );
                }
                addDocument(txn, docChanges, id, op[1]);
                break;
              }
              case import_diff.RecordOpType.Patch: {
                patchDocument(txn, docChanges, id, op[1]);
                break;
              }
              case import_diff.RecordOpType.Remove: {
                const doc = txn.get(id);
                if (!doc) {
                  continue;
                }
                txn.delete(id);
                propagateOp(docChanges, id, op, doc, void 0);
                break;
              }
            }
          }
        }
        return { docChanges, presenceChanges };
      },
      { id: this.internalTxnId, emitChanges: "when-different" }
    );
    this.lastDocumentClock = documentClock;
    let pushResult;
    if (changes && session) {
      result.docChanges.diffs = { networkDiff: (0, import_TLSyncStorage.toNetworkDiff)(changes) ?? {}, diff: changes };
    }
    if ((0, import_utils.isEqual)(result.docChanges.diffs?.networkDiff, message.diff)) {
      pushResult = {
        type: "push_result",
        clientClock: message.clientClock,
        serverClock: documentClock,
        action: "commit"
      };
    } else if (!result.docChanges.diffs?.networkDiff) {
      pushResult = {
        type: "push_result",
        clientClock: message.clientClock,
        serverClock: documentClock,
        action: "discard"
      };
    } else if (session) {
      const diff = this.migrateDiffOrRejectSession(
        session.sessionId,
        session.serializedSchema,
        session.requiresDownMigrations,
        result.docChanges.diffs.diff,
        result.docChanges.diffs.networkDiff
      );
      if (diff.ok) {
        pushResult = {
          type: "push_result",
          clientClock: message.clientClock,
          serverClock: documentClock,
          action: { rebaseWithDiff: diff.value }
        };
      }
    }
    if (session && pushResult) {
      this._unsafe_sendMessage(session.sessionId, pushResult);
    }
    if (result.docChanges.diffs || result.presenceChanges.diffs) {
      this.broadcastPatch(
        {
          puts: {
            ...result.docChanges.diffs?.diff.puts,
            ...result.presenceChanges.diffs?.diff.puts
          },
          deletes: [
            ...result.docChanges.diffs?.diff.deletes ?? [],
            ...result.presenceChanges.diffs?.diff.deletes ?? []
          ]
        },
        {
          ...result.docChanges.diffs?.networkDiff,
          ...result.presenceChanges.diffs?.networkDiff
        },
        session?.sessionId
      );
    }
    if (result.presenceChanges.diffs) {
      queueMicrotask(() => {
        this.onPresenceChange?.();
      });
    }
  }
  /**
   * Handle the event when a client disconnects. Cleans up the session and
   * removes any presence information.
   *
   * @param sessionId - The session that disconnected
   * @example
   * ```ts
   * websocket.onClose(() => {
   *   room.handleClose(sessionId)
   * })
   * ```
   */
  handleClose(sessionId) {
    this.cancelSession(sessionId);
  }
}
class PresenceStore {
  presences = new import_store.AtomMap("presences");
  get(id) {
    return this.presences.get(id);
  }
  set(id, state) {
    this.presences.set(id, state);
  }
  delete(id) {
    this.presences.delete(id);
  }
  values() {
    return this.presences.values();
  }
}
