// @tldraw/sync-core@5.2.5/dist-cjs/lib/TLSyncStorage.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var TLSyncStorage_exports = {};
__export(TLSyncStorage_exports, {
  convertStoreSnapshotToRoomSnapshot: () => convertStoreSnapshotToRoomSnapshot,
  loadSnapshotIntoStorage: () => loadSnapshotIntoStorage,
  toNetworkDiff: () => toNetworkDiff
});
module.exports = __toCommonJS(TLSyncStorage_exports);
var import_utils = require("@tldraw/utils");
var import_diff = require("./diff");
function toNetworkDiff(diff) {
  const networkDiff = {};
  for (const [id, put] of (0, import_utils.objectMapEntriesIterable)(diff.puts)) {
    if (Array.isArray(put)) {
      const patch = (0, import_diff.diffRecord)(put[0], put[1]);
      if (patch) {
        networkDiff[id] = [import_diff.RecordOpType.Patch, patch];
      }
    } else {
      networkDiff[id] = [import_diff.RecordOpType.Put, put];
    }
  }
  for (const id of diff.deletes) {
    networkDiff[id] = [import_diff.RecordOpType.Remove];
  }
  return networkDiff;
}
function loadSnapshotIntoStorage(txn, schema, snapshot) {
  snapshot = convertStoreSnapshotToRoomSnapshot(snapshot);
  (0, import_utils.assert)(snapshot.schema, "Schema is required");
  const docIds = /* @__PURE__ */ new Set();
  for (const doc of snapshot.documents) {
    docIds.add(doc.state.id);
    const existing = txn.get(doc.state.id);
    if ((0, import_utils.isEqual)(existing, doc.state)) continue;
    txn.set(doc.state.id, doc.state);
  }
  for (const id of txn.keys()) {
    if (!docIds.has(id)) {
      txn.delete(id);
    }
  }
  txn.setSchema(snapshot.schema);
  schema.migrateStorage(txn);
}
function convertStoreSnapshotToRoomSnapshot(snapshot) {
  if ("documents" in snapshot) return snapshot;
  return {
    clock: 0,
    documentClock: 0,
    documents: (0, import_utils.objectMapValues)(snapshot.store).map((state) => ({
      state,
      lastChangedClock: 0
    })),
    schema: snapshot.schema,
    tombstones: {}
  };
}
