// @tldraw/sync@5.2.5/dist-esm/useSync.mjs
import { atom, transact } from "@tldraw/state";
import {
  ClientWebSocketAdapter,
  TLRemoteSyncError,
  TLSyncClient,
  TLSyncErrorCloseEventReason
} from "@tldraw/sync-core";
import { useEffect } from "react";
import {
  TAB_ID,
  UserRecordType,
  computed,
  createCachedUserResolve,
  createPresenceStateDerivation,
  registerColorsFromThemes,
  registerFontsFromThemes,
  resolveThemes,
  createTLStore,
  createUserId,
  defaultUserPreferences,
  defaultUserStore,
  getDefaultUserPresence,
  getUserPreferences,
  uniqueId,
  useEvent,
  useReactiveEvent,
  useRefState,
  useTLSchemaFromUtils,
  useValue
} from "tldraw";
const MULTIPLAYER_EVENT_NAME = "multiplayer.client";
const defaultCustomMessageHandler = () => {
};
function useSync(opts) {
  const [state, setState] = useRefState(null);
  const {
    uri,
    roomId = "default",
    assets,
    users: _users,
    onMount,
    connect,
    trackAnalyticsEvent: track,
    getUserPresence: _getUserPresence,
    onCustomMessageReceived: _onCustomMessageReceived,
    themes,
    ...schemaOpts
  } = opts;
  const __never__ = 0;
  const resolvedThemes = resolveThemes(themes);
  registerColorsFromThemes(resolvedThemes);
  registerFontsFromThemes(resolvedThemes);
  const schema = useTLSchemaFromUtils(schemaOpts);
  const getUserPresence = useReactiveEvent(
    _getUserPresence ?? getDefaultUserPresence
  );
  const onCustomMessageReceived = useEvent(_onCustomMessageReceived ?? defaultCustomMessageHandler);
  useEffect(() => {
    const storeId = uniqueId();
    const users = _users ? {
      currentUser: _users.currentUser,
      resolve: _users.resolve ?? createCachedUserResolve((userId) => {
        const current = _users.currentUser.get();
        return current && current.id === createUserId(userId) ? current : null;
      })
    } : {
      currentUser: defaultUserStore.currentUser,
      resolve: createCachedUserResolve((userId) => {
        const current = defaultUserStore.currentUser.get();
        if (current && current.id === createUserId(userId)) return current;
        const presences = store.query.records("instance_presence").get();
        const match = presences.find((p) => p.userId === createUserId(userId));
        if (match) {
          return UserRecordType.create({
            id: createUserId(userId),
            name: match.userName,
            color: match.color
          });
        }
        return null;
      })
    };
    const currentUser = computed("currentUser", () => {
      const user = users.currentUser.get();
      if (user) return user;
      const prefs = getUserPreferences();
      return UserRecordType.create({
        id: createUserId(prefs.id),
        name: prefs.name ?? "",
        color: prefs.color ?? defaultUserPreferences.color
      });
    });
    let socket;
    if (connect) {
      if (uri) {
        throw new Error("uri and connect cannot be used together");
      }
      socket = connect({
        sessionId: TAB_ID,
        storeId
      });
    } else if (uri) {
      if (connect) {
        throw new Error("uri and connect cannot be used together");
      }
      socket = new ClientWebSocketAdapter(async () => {
        const uriString = typeof uri === "string" ? uri : await uri();
        const withParams = new URL(uriString);
        if (withParams.searchParams.has("sessionId")) {
          throw new Error(
            'useSync. "sessionId" is a reserved query param name. Please use a different name'
          );
        }
        if (withParams.searchParams.has("storeId")) {
          throw new Error(
            'useSync. "storeId" is a reserved query param name. Please use a different name'
          );
        }
        withParams.searchParams.set("sessionId", TAB_ID);
        withParams.searchParams.set("storeId", storeId);
        return withParams.toString();
      });
    } else {
      throw new Error("uri or connect must be provided");
    }
    let didCancel = false;
    function getConnectionStatus() {
      return socket.connectionStatus === "error" ? "offline" : socket.connectionStatus;
    }
    const collaborationStatusSignal = atom("collaboration status", getConnectionStatus());
    const unsubscribeFromConnectionStatus = socket.onStatusChange(() => {
      collaborationStatusSignal.set(getConnectionStatus());
    });
    const syncMode = atom("sync mode", "readwrite");
    const store = createTLStore({
      id: storeId,
      schema,
      assets,
      users,
      onMount,
      collaboration: {
        status: collaborationStatusSignal,
        mode: syncMode
      }
    });
    const presence = createPresenceStateDerivation(currentUser, {
      getUserPresence
    })(store);
    const otherSessions = store.query.ids("instance_presence");
    const presenceMode = computed("presenceMode", () => {
      return otherSessions.get().size === 0 ? "solo" : "full";
    });
    const client = new TLSyncClient({
      store,
      socket,
      didCancel: () => didCancel,
      onLoad(client2) {
        track?.(MULTIPLAYER_EVENT_NAME, { name: "load", roomId });
        setState({ readyClient: client2 });
      },
      onSyncError(reason) {
        console.error("sync error", reason);
        switch (reason) {
          case TLSyncErrorCloseEventReason.NOT_FOUND:
            track?.(MULTIPLAYER_EVENT_NAME, { name: "room-not-found", roomId });
            break;
          case TLSyncErrorCloseEventReason.FORBIDDEN:
            track?.(MULTIPLAYER_EVENT_NAME, { name: "forbidden", roomId });
            break;
          case TLSyncErrorCloseEventReason.NOT_AUTHENTICATED:
            track?.(MULTIPLAYER_EVENT_NAME, { name: "not-authenticated", roomId });
            break;
          case TLSyncErrorCloseEventReason.RATE_LIMITED:
            track?.(MULTIPLAYER_EVENT_NAME, { name: "rate-limited", roomId });
            break;
          default:
            track?.(MULTIPLAYER_EVENT_NAME, { name: "sync-error:" + reason, roomId });
            break;
        }
        setState({ error: new TLRemoteSyncError(reason) });
        socket.close();
      },
      onAfterConnect(_, { isReadonly }) {
        transact(() => {
          syncMode.set(isReadonly ? "readonly" : "readwrite");
          store.ensureStoreIsUsable();
        });
      },
      onCustomMessageReceived,
      presence,
      presenceMode
    });
    return () => {
      didCancel = true;
      unsubscribeFromConnectionStatus();
      client.close();
      socket.close();
    };
  }, [
    assets,
    onMount,
    connect,
    _users,
    roomId,
    schema,
    setState,
    track,
    uri,
    getUserPresence,
    onCustomMessageReceived
  ]);
  return useValue(
    "remote synced store",
    () => {
      if (!state) return { status: "loading" };
      if (state.error) return { status: "error", error: state.error };
      if (!state.readyClient) return { status: "loading" };
      const connectionStatus = state.readyClient.socket.connectionStatus;
      return {
        status: "synced-remote",
        connectionStatus: connectionStatus === "error" ? "offline" : connectionStatus,
        store: state.readyClient.store
      };
    },
    [state]
  );
}
export {
  useSync
};
