// @apm-js-collab/code-transformer-bundler-plugins@0.6.2/dist/cjs/webpack.cjs
const require_instrumentation_serde = require("./instrumentation-serde-BFky_HG6.cjs");
let path = require("path");
let url = require("url");
//#region src/webpack.ts
var LOADER_PATH = (0, path.resolve)((0, path.dirname)((0, url.fileURLToPath)(require("url").pathToFileURL(__filename).href)), "..", "cjs", "webpack-loader.cjs");
var DIAGNOSTICS_STATE_KEY = "__codeTransformerWebpackDiagnostics";
/**
* Asset names of the chunk holding each entry module. Deliberately not
* `entrypoint.getFiles()`, which also lists the initial chunks an entry
* depends on, and not `compilation.getAssets()`, which lists async chunks too.
*/
function entryAssetNames(compilation) {
	const names = /* @__PURE__ */ new Set();
	for (const entrypoint of compilation.entrypoints.values()) {
		const chunk = entrypoint.getEntrypointChunk?.();
		for (const file of chunk?.files ?? []) names.add(file);
	}
	return names;
}
var CodeTransformerWebpackPlugin = class {
	constructor(options) {
		this.options = options;
	}
	apply(compiler) {
		const webpack = compiler.webpack;
		compiler.options.module = compiler.options.module || { rules: [] };
		compiler.options.module.rules = compiler.options.module.rules || [];
		compiler.options.module.rules.unshift({
			test: /\.(c|m)?jsx?$|\.tsx?$/,
			enforce: "pre",
			use: [{
				loader: LOADER_PATH,
				options: {
					instrumentations: require_instrumentation_serde.serializeInstrumentations(this.options.instrumentations),
					...this.options.dcModule ? { dcModule: this.options.dcModule } : {}
				}
			}]
		});
		if (this.options.injectDiagnostics) {
			const ConcatSource = webpack?.sources?.ConcatSource;
			if (ConcatSource && webpack?.Compilation) compiler.hooks.thisCompilation.tap("code-transformer", (compilation) => {
				compilation[DIAGNOSTICS_STATE_KEY] = {
					transformedModules: /* @__PURE__ */ new Set(),
					failedModules: /* @__PURE__ */ new Set()
				};
				compilation.hooks.processAssets.tap({
					name: "code-transformer",
					stage: webpack.Compilation.PROCESS_ASSETS_STAGE_SUMMARIZE
				}, () => {
					const state = compilation[DIAGNOSTICS_STATE_KEY];
					if (!state) return;
					const injectCode = this.options.injectDiagnostics?.({
						transformedModules: Array.from(state.transformedModules),
						failedModules: Array.from(state.failedModules)
					});
					if (!injectCode) return;
					for (const assetName of entryAssetNames(compilation)) {
						if (!/\.(js|ts|jsx|tsx|mjs|cjs)(\?[^?]*)?(#[^#]*)?$/.test(assetName)) continue;
						compilation.updateAsset(assetName, (source) => new ConcatSource(injectCode, source));
					}
				});
			});
		}
	}
};
function codeTransformerWebpack(options) {
	return new CodeTransformerWebpackPlugin(options);
}
//#endregion
module.exports = codeTransformerWebpack;
