// @tldraw/sync-core@5.2.5/dist-cjs/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var index_exports = {};
__export(index_exports, {
  ClientWebSocketAdapter: () => import_ClientWebSocketAdapter.ClientWebSocketAdapter,
  DEFAULT_INITIAL_SNAPSHOT: () => import_InMemorySyncStorage.DEFAULT_INITIAL_SNAPSHOT,
  DurableObjectSqliteSyncWrapper: () => import_DurableObjectSqliteSyncWrapper.DurableObjectSqliteSyncWrapper,
  InMemorySyncStorage: () => import_InMemorySyncStorage.InMemorySyncStorage,
  JsonChunkAssembler: () => import_chunk.JsonChunkAssembler,
  NodeSqliteWrapper: () => import_NodeSqliteWrapper.NodeSqliteWrapper,
  ReconnectManager: () => import_ClientWebSocketAdapter.ReconnectManager,
  RecordOpType: () => import_diff.RecordOpType,
  RoomSessionState: () => import_RoomSession.RoomSessionState,
  SQLiteSyncStorage: () => import_SQLiteSyncStorage.SQLiteSyncStorage,
  TLIncompatibilityReason: () => import_protocol.TLIncompatibilityReason,
  TLRemoteSyncError: () => import_TLRemoteSyncError.TLRemoteSyncError,
  TLSocketRoom: () => import_TLSocketRoom.TLSocketRoom,
  TLSyncClient: () => import_TLSyncClient.TLSyncClient,
  TLSyncErrorCloseEventCode: () => import_TLSyncClient.TLSyncErrorCloseEventCode,
  TLSyncErrorCloseEventReason: () => import_TLSyncClient.TLSyncErrorCloseEventReason,
  TLSyncRoom: () => import_TLSyncRoom.TLSyncRoom,
  ValueOpType: () => import_diff.ValueOpType,
  applyObjectDiff: () => import_diff.applyObjectDiff,
  chunk: () => import_chunk.chunk,
  diffRecord: () => import_diff.diffRecord,
  getNetworkDiff: () => import_diff.getNetworkDiff,
  getTlsyncProtocolVersion: () => import_protocol.getTlsyncProtocolVersion,
  loadSnapshotIntoStorage: () => import_TLSyncStorage.loadSnapshotIntoStorage
});
module.exports = __toCommonJS(index_exports);
var import_utils = require("@tldraw/utils");
var import_chunk = require("./lib/chunk");
var import_ClientWebSocketAdapter = require("./lib/ClientWebSocketAdapter");
var import_diff = require("./lib/diff");
var import_DurableObjectSqliteSyncWrapper = require("./lib/DurableObjectSqliteSyncWrapper");
var import_InMemorySyncStorage = require("./lib/InMemorySyncStorage");
var import_NodeSqliteWrapper = require("./lib/NodeSqliteWrapper");
var import_protocol = require("./lib/protocol");
var import_RoomSession = require("./lib/RoomSession");
var import_SQLiteSyncStorage = require("./lib/SQLiteSyncStorage");
var import_TLRemoteSyncError = require("./lib/TLRemoteSyncError");
var import_TLSocketRoom = require("./lib/TLSocketRoom");
var import_TLSyncClient = require("./lib/TLSyncClient");
var import_TLSyncRoom = require("./lib/TLSyncRoom");
var import_TLSyncStorage = require("./lib/TLSyncStorage");
(0, import_utils.registerTldrawLibraryVersion)(
  "@tldraw/sync-core",
  "5.2.5",
  "cjs"
);
