// @tldraw/sync@5.2.5/dist-cjs/useSync.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var useSync_exports = {};
__export(useSync_exports, {
  useSync: () => useSync
});
module.exports = __toCommonJS(useSync_exports);
var import_state = require("@tldraw/state");
var import_sync_core = require("@tldraw/sync-core");
var import_react = require("react");
var import_tldraw = require("tldraw");
const MULTIPLAYER_EVENT_NAME = "multiplayer.client";
const defaultCustomMessageHandler = () => {
};
function useSync(opts) {
  const [state, setState] = (0, import_tldraw.useRefState)(null);
  const {
    uri,
    roomId = "default",
    assets,
    users: _users,
    onMount,
    connect,
    trackAnalyticsEvent: track,
    getUserPresence: _getUserPresence,
    onCustomMessageReceived: _onCustomMessageReceived,
    themes,
    ...schemaOpts
  } = opts;
  const __never__ = 0;
  const resolvedThemes = (0, import_tldraw.resolveThemes)(themes);
  (0, import_tldraw.registerColorsFromThemes)(resolvedThemes);
  (0, import_tldraw.registerFontsFromThemes)(resolvedThemes);
  const schema = (0, import_tldraw.useTLSchemaFromUtils)(schemaOpts);
  const getUserPresence = (0, import_tldraw.useReactiveEvent)(
    _getUserPresence ?? import_tldraw.getDefaultUserPresence
  );
  const onCustomMessageReceived = (0, import_tldraw.useEvent)(_onCustomMessageReceived ?? defaultCustomMessageHandler);
  (0, import_react.useEffect)(() => {
    const storeId = (0, import_tldraw.uniqueId)();
    const users = _users ? {
      currentUser: _users.currentUser,
      resolve: _users.resolve ?? (0, import_tldraw.createCachedUserResolve)((userId) => {
        const current = _users.currentUser.get();
        return current && current.id === (0, import_tldraw.createUserId)(userId) ? current : null;
      })
    } : {
      currentUser: import_tldraw.defaultUserStore.currentUser,
      resolve: (0, import_tldraw.createCachedUserResolve)((userId) => {
        const current = import_tldraw.defaultUserStore.currentUser.get();
        if (current && current.id === (0, import_tldraw.createUserId)(userId)) return current;
        const presences = store.query.records("instance_presence").get();
        const match = presences.find((p) => p.userId === (0, import_tldraw.createUserId)(userId));
        if (match) {
          return import_tldraw.UserRecordType.create({
            id: (0, import_tldraw.createUserId)(userId),
            name: match.userName,
            color: match.color
          });
        }
        return null;
      })
    };
    const currentUser = (0, import_tldraw.computed)("currentUser", () => {
      const user = users.currentUser.get();
      if (user) return user;
      const prefs = (0, import_tldraw.getUserPreferences)();
      return import_tldraw.UserRecordType.create({
        id: (0, import_tldraw.createUserId)(prefs.id),
        name: prefs.name ?? "",
        color: prefs.color ?? import_tldraw.defaultUserPreferences.color
      });
    });
    let socket;
    if (connect) {
      if (uri) {
        throw new Error("uri and connect cannot be used together");
      }
      socket = connect({
        sessionId: import_tldraw.TAB_ID,
        storeId
      });
    } else if (uri) {
      if (connect) {
        throw new Error("uri and connect cannot be used together");
      }
      socket = new import_sync_core.ClientWebSocketAdapter(async () => {
        const uriString = typeof uri === "string" ? uri : await uri();
        const withParams = new URL(uriString);
        if (withParams.searchParams.has("sessionId")) {
          throw new Error(
            'useSync. "sessionId" is a reserved query param name. Please use a different name'
          );
        }
        if (withParams.searchParams.has("storeId")) {
          throw new Error(
            'useSync. "storeId" is a reserved query param name. Please use a different name'
          );
        }
        withParams.searchParams.set("sessionId", import_tldraw.TAB_ID);
        withParams.searchParams.set("storeId", storeId);
        return withParams.toString();
      });
    } else {
      throw new Error("uri or connect must be provided");
    }
    let didCancel = false;
    function getConnectionStatus() {
      return socket.connectionStatus === "error" ? "offline" : socket.connectionStatus;
    }
    const collaborationStatusSignal = (0, import_state.atom)("collaboration status", getConnectionStatus());
    const unsubscribeFromConnectionStatus = socket.onStatusChange(() => {
      collaborationStatusSignal.set(getConnectionStatus());
    });
    const syncMode = (0, import_state.atom)("sync mode", "readwrite");
    const store = (0, import_tldraw.createTLStore)({
      id: storeId,
      schema,
      assets,
      users,
      onMount,
      collaboration: {
        status: collaborationStatusSignal,
        mode: syncMode
      }
    });
    const presence = (0, import_tldraw.createPresenceStateDerivation)(currentUser, {
      getUserPresence
    })(store);
    const otherSessions = store.query.ids("instance_presence");
    const presenceMode = (0, import_tldraw.computed)("presenceMode", () => {
      return otherSessions.get().size === 0 ? "solo" : "full";
    });
    const client = new import_sync_core.TLSyncClient({
      store,
      socket,
      didCancel: () => didCancel,
      onLoad(client2) {
        track?.(MULTIPLAYER_EVENT_NAME, { name: "load", roomId });
        setState({ readyClient: client2 });
      },
      onSyncError(reason) {
        console.error("sync error", reason);
        switch (reason) {
          case import_sync_core.TLSyncErrorCloseEventReason.NOT_FOUND:
            track?.(MULTIPLAYER_EVENT_NAME, { name: "room-not-found", roomId });
            break;
          case import_sync_core.TLSyncErrorCloseEventReason.FORBIDDEN:
            track?.(MULTIPLAYER_EVENT_NAME, { name: "forbidden", roomId });
            break;
          case import_sync_core.TLSyncErrorCloseEventReason.NOT_AUTHENTICATED:
            track?.(MULTIPLAYER_EVENT_NAME, { name: "not-authenticated", roomId });
            break;
          case import_sync_core.TLSyncErrorCloseEventReason.RATE_LIMITED:
            track?.(MULTIPLAYER_EVENT_NAME, { name: "rate-limited", roomId });
            break;
          default:
            track?.(MULTIPLAYER_EVENT_NAME, { name: "sync-error:" + reason, roomId });
            break;
        }
        setState({ error: new import_sync_core.TLRemoteSyncError(reason) });
        socket.close();
      },
      onAfterConnect(_, { isReadonly }) {
        (0, import_state.transact)(() => {
          syncMode.set(isReadonly ? "readonly" : "readwrite");
          store.ensureStoreIsUsable();
        });
      },
      onCustomMessageReceived,
      presence,
      presenceMode
    });
    return () => {
      didCancel = true;
      unsubscribeFromConnectionStatus();
      client.close();
      socket.close();
    };
  }, [
    assets,
    onMount,
    connect,
    _users,
    roomId,
    schema,
    setState,
    track,
    uri,
    getUserPresence,
    onCustomMessageReceived
  ]);
  return (0, import_tldraw.useValue)(
    "remote synced store",
    () => {
      if (!state) return { status: "loading" };
      if (state.error) return { status: "error", error: state.error };
      if (!state.readyClient) return { status: "loading" };
      const connectionStatus = state.readyClient.socket.connectionStatus;
      return {
        status: "synced-remote",
        connectionStatus: connectionStatus === "error" ? "offline" : connectionStatus,
        store: state.readyClient.store
      };
    },
    [state]
  );
}
